// cloud-firestore/src/main/storage/storage.ts
import admin from 'firebase-admin';
import { Observable } from 'rxjs';
import { Transaction } from '../firestore/transaction';
export class Storage {
    observer:Observable<any> = new Observable<any>(observer => {            
        observer.next();
        observer.complete();
    });
    database:any;
    onInit(config:any){
        if(!config)
            return;
        let adminConfig = {
            credential: admin.credential.cert(config.credential),
            storageBucket: config.server
            };
        if(config.appID){
            admin.initializeApp(adminConfig, config.appID);
            let app = admin.app(config.appID);
            this.database = admin.firestore(app);
            
        }else{
            throw new Error("AppID required.");       
        }
    }
    connect(isTransaction:boolean, currentTrans): Observable<any> {        
        let observable = new Observable<any>(observer => {
            if((!isTransaction && (currentTrans && currentTrans.type === Transaction.NO_TRANSACTION)) 
            || (currentTrans && currentTrans.type === Transaction.REQUIRED_TRANSACTION)){
                observer.next(currentTrans);
                observer.complete();
            }else{
                if(isTransaction){
                        observer.next(new Transaction({}, Transaction.REQUIRED_TRANSACTION));
                        observer.complete();                    
                }else{
                    observer.next(new Transaction({}, Transaction.NO_TRANSACTION));
                    observer.complete();
                }
            }
        });
        return observable;
    }
    commit(transaction:any){
        if(transaction)
            return transaction.commit();
    }
    rollback(transaction:any){
        if(transaction)
            return transaction.rollback();
    }
    post(input:any, pConfig:any, transaction:any) {
        return new Observable<any>(observer => { 
            let file = input.files[0];
            var bucket = admin.storage().bucket();
            bucket.upload(file.name, {destination : file.name}).then((data) => {
                let file = data[0];
                observer.next([{downloadUrl: "https://firebasestorage.googleapis.com/v0/b/" + bucket.name + "/o/" + encodeURIComponent(file.name) + "?alt=media&token=" + this.uuidv4()}, transaction]);
                observer.complete();              
            }).catch(err=>{
                observer.error(err); 
            });  
                     
        });
    }
    uuidv4() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
          var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
          return v.toString(16);
        });
      }
    put(input:any, pConfig:any, transaction:any) {
        return new Observable<any>(observer => { 
            
        });
    }
    delete(input:any, pConfig:any, transaction:any) {
        return new Observable<any>(observer => { 
            
        });
    }
    get(input:any, pConfig:any, transaction:any) {
        return new Observable<any>(observer => { 
           
        });
    }
}
