// quiver-component/out/lib/abstract.js
"use strict";

var _interopRequire = function (obj) { return obj && obj.__esModule ? obj["default"] : obj; };

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var assign = _interopRequire(require("object.assign"));

var _extensibleComponent = require("./extensible-component");

var ExtensibleHandler = _extensibleComponent.ExtensibleHandler;
var ExtensibleMiddleware = _extensibleComponent.ExtensibleMiddleware;

const _componentKey = Symbol("_componentKey");

const defineAbstractComponent = function (Parent, mixin) {
  let AbstractComponent = (function (_Parent) {
    function AbstractComponent(componentKey) {
      let options = arguments[1] === undefined ? {} : arguments[1];

      _classCallCheck(this, AbstractComponent);

      _get(Object.getPrototypeOf(AbstractComponent.prototype), "constructor", this).call(this, options);
      this[_componentKey] = componentKey;
    }

    _inherits(AbstractComponent, _Parent);

    _createClass(AbstractComponent, {
      implement: {
        value: function implement(componentMap) {
          if (!this.concreteComponent) {
            const componentKey = this[_componentKey];
            const concreteComponent = componentMap[componentKey];

            if (concreteComponent) {
              this.validateConcreteComponent(concreteComponent);
              this.subComponents.concreteComponent = concreteComponent;
            }
          }

          return _get(Object.getPrototypeOf(AbstractComponent.prototype), "implement", this).call(this, componentMap);
        }
      },
      concreteComponent: {
        get: function () {
          return this.subComponents.concreteComponent;
        }
      }
    });

    return AbstractComponent;
  })(Parent);

  assign(AbstractComponent.prototype, mixin);

  return AbstractComponent;
};

const AbstractHandler = defineAbstractComponent(ExtensibleHandler, {
  toMainHandleableBuilder: function toMainHandleableBuilder() {
    var _ref = this;

    const concreteComponent = _ref.concreteComponent;

    if (!concreteComponent) {
      throw new Error("Abstract handler component " + "not implemented: " + this[_componentKey]);
    }

    return concreteComponent.toHandleableBuilder();
  },

  validateConcreteComponent: function validateConcreteComponent(component) {
    if (!component.isHandlerComponent) {
      throw new Error("Concrete component in " + "implementation map is not handler component: " + componentKey);
    }
  }
});

exports.AbstractHandler = AbstractHandler;
const AbstractMiddleware = defineAbstractComponent(ExtensibleMiddleware, {
  toMainHandleableMiddleware: function toMainHandleableMiddleware() {
    var _ref = this;

    const concreteComponent = _ref.concreteComponent;

    if (!concreteComponent) throw new Error("Abstract middleware component " + "not implemented: " + this[_componentKey]);

    return concreteComponent.toMainHandleableMiddleware();
  },

  validateConcreteComponent: function validateConcreteComponent(component) {
    if (!component.isMiddlewareComponent) {
      throw new Error("Concrete component in " + "implementation map is not middleware component: " + componentKey);
    }
  }
});

exports.AbstractMiddleware = AbstractMiddleware;
const abstractHandler = function (componentKey) {
  return new AbstractHandler(componentKey);
};

exports.abstractHandler = abstractHandler;
const abstractMiddleware = function (componentKey) {
  return new AbstractMiddleware(componentKey);
};
exports.abstractMiddleware = abstractMiddleware;
