// quiver-component/out/lib/bundle.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var async = require("quiver-promise").async;

var simpleToStreamHandler = require("quiver-simple-handler").simpleToStreamHandler;

var extendHandler = require("./extend").extendHandler;

var HandleableBuilder = require("./handleable-builder").HandleableBuilder;

var handleableMiddleware = require("./handleable-middleware").handleableMiddleware;

var _utilLoader = require("./util/loader");

var loadStreamHandler = _utilLoader.loadStreamHandler;
var simpleHandlerLoader = _utilLoader.simpleHandlerLoader;

const _bundleFields = Symbol("bundleFields");

const streamHandlerConverter = function (handler) {
  return {
    streamHandler: safePromised(handler)
  };
};

const simpleHandlerConverter = function (inType, outType) {
  return function (handler) {
    return {
      streamHandler: simpleToStreamHandler(handler, inType, outType)
    };
  };
};

const bundleFieldMiddleware = function (field, handlerConverter) {
  return handleableMiddleware(async(function* (config, builder) {
    const bundle = yield builder(config);
    const handler = bundle[field];

    if (!handler) {
      throw new Error("missing handler field in bundle: " + field);
    }

    return handlerConverter(handler);
  }));
};

const bundleFieldComponent = function (bundleComponent, field, handlerConverter, handlerLoader) {
  return extendHandler(bundleComponent).middleware(bundleFieldMiddleware(field, handlerConverter)).setLoader(handlerLoader);
};

let HandlerBundle = exports.HandlerBundle = (function (_HandleableBuilder) {
  function HandlerBundle(bundleBuilder) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, HandlerBundle);

    _get(Object.getPrototypeOf(HandlerBundle.prototype), "constructor", this).call(this, bundleBuilder, options);
    this[_bundleFields] = {};
  }

  _inherits(HandlerBundle, _HandleableBuilder);

  _createClass(HandlerBundle, {
    toHandlerComponents: {
      value: function toHandlerComponents() {
        const handlerComponents = {};
        const bundleFields = this[_bundleFields];

        for (let field in bundleFields) {
          var _bundleFields$field = bundleFields[field];
          const handlerConverter = _bundleFields$field.handlerConverter;
          const handlerLoader = _bundleFields$field.handlerLoader;

          handlerComponents[field] = bundleFieldComponent(this, field, handlerConverter, handlerLoader);
        }

        return handlerComponents;
      }
    },
    bundleField: {
      value: function bundleField(handlerName, handlerConverter, handlerLoader) {
        const bundleFields = this[_bundleFields];

        if (bundleFields[handlerName]) throw new Error("bundle field is already defined: " + handlerName);

        bundleFields[handlerName] = { handlerConverter: handlerConverter, handlerLoader: handlerLoader };

        return this;
      }
    },
    streamHandler: {
      value: function streamHandler(handlerName) {
        return this.bundleField(handlerName, streamHandlerConverter, loadStreamHandler);
      }
    },
    simpleHandler: {
      value: function simpleHandler(handlerName, inType, outType) {
        return this.bundleField(handlerName, simpleHandlerConverter(inType, outType), simpleHandlerLoader(inType, outType));
      }
    }
  });

  return HandlerBundle;
})(HandleableBuilder);

const handlerBundle = function (bundleBuilder) {
  return new HandlerBundle(bundleBuilder);
};
exports.handlerBundle = handlerBundle;
