// quiver-component/out/lib/component.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});
const assertComponent = function (component) {
  if (!component.isQuiverComponent) {
    throw new Error("object must be of type Component");
  }
};

// Random ID for easier identifying
const randomId = function () {
  return Symbol((Math.random() * 268435456 | 0).toString(16));
};

let Component = exports.Component = (function () {
  function Component() {
    let options = arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, Component);

    const name = options.name;
    var _options$subComponents = options.subComponents;
    const subComponents = _options$subComponents === undefined ? {} : _options$subComponents;

    this._id = randomId();
    this._options = options;

    if (name) {
      this.name = name;
    }

    for (let key in subComponents) {
      const component = subComponents[key];
      assertComponent(component);
    }

    this._subComponents = subComponents;
  }

  _createClass(Component, {
    id: {
      get: function () {
        return this._id;
      }
    },
    type: {
      get: function () {
        return "component";
      }
    },
    isQuiverComponent: {
      get: function () {
        return true;
      }
    },
    subComponents: {
      get: function () {
        return this._subComponents;
      }
    },
    addSubComponent: {
      value: function addSubComponent(key, component) {
        if (!component.isQuiverComponent) {
          throw new Error("Subcomponent must be Quiver component");
        }

        var _ref = this;

        const subComponents = _ref.subComponents;

        if (subComponents[key]) {
          throw new Error("Subcomponent already registered at given key");
        }

        subComponents[key] = component;
      }
    },
    getSubComponent: {
      value: function getSubComponent(key) {
        return this.subComponents[key];
      }
    },
    setName: {
      value: function setName(name) {
        this.name = name;
        return this;
      }
    },
    clone: {
      value: function clone() {
        const newInstance = Object.create(this);

        const privateId = randomId();
        Object.defineProperty(newInstance, "id", {
          get: function get() {
            return privateId;
          }
        });

        return newInstance;
      }
    },
    applyMap: {
      value: function applyMap(mapper) {
        let mapTable = arguments[1] === undefined ? {} : arguments[1];

        const currentId = this.id;

        if (mapTable[currentId]) {
          return mapTable[currentId];
        }

        return mapTable[currentId] = mapper(this, mapTable);
      }
    },
    map: {
      value: function map(mapper) {
        let mapTable = arguments[1] === undefined ? {} : arguments[1];

        const clone = this.clone();
        this.doMap(clone, mapper, mapTable);
        return clone;
      }
    },
    each: {
      value: function each(iteratee) {
        var _ref = this;

        const subComponents = _ref.subComponents;

        for (let key in subComponents) {
          iteratee(subComponents[key]);
        }
      }
    },
    doMap: {
      value: function doMap(target, mapper, mapTable) {
        var _ref = this;

        const subComponents = _ref.subComponents;

        const newSubComponents = {};

        for (let key in subComponents) {
          const component = subComponents[key];

          newSubComponents[key] = component.applyMap(mapper, mapTable);
        }

        target._subComponents = newSubComponents;
      }
    },
    factory: {
      value: function factory() {
        var _this = this;

        return function () {
          let forkTable = arguments[0] === undefined ? {} : arguments[0];
          return _this.fork(forkTable);
        };
      }
    },
    fork: {
      value: function fork() {
        let forkTable = arguments[0] === undefined ? {} : arguments[0];

        return this.applyMap(function (component, mapTable) {
          return component.map(function (subComponent, mapTable) {
            return subComponent.fork(mapTable);
          }, mapTable);
        }, forkTable);
      }
    },
    implement: {
      value: function implement(componentMap) {
        this.each(function (component) {
          return component.implement(componentMap);
        });

        return this;
      }
    },
    toJson: {
      value: function toJson() {
        const json = {
          id: this.id.toString(),
          type: this.type
        };

        if (this.name) {
          json.name = this.name;
        }

        return json;
      }
    },
    toString: {
      value: function toString() {
        return JSON.stringify(this.toJson(), undefined, 2);
      }
    },
    inspect: {
      value: function inspect() {
        return this.toString();
      }
    }
  });

  return Component;
})();
