// quiver-component/out/lib/extensible-component.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var copy = require("quiver-object").copy;

var Component = require("./component").Component;

var listComponent = require("./composite/list").listComponent;

var _utilMiddleware = require("./util/middleware");

var combineMiddlewares = _utilMiddleware.combineMiddlewares;
var combineMiddlewareComponents = _utilMiddleware.combineMiddlewareComponents;
var combineBuilderWithMiddleware = _utilMiddleware.combineBuilderWithMiddleware;

var loadHandleable = require("./util/loader").loadHandleable;

const _middlewareComponents = Symbol("_middlewareComponents");

const copyConfigBuilder = function (builder) {
  return function (config) {
    return builder(copy(config));
  };
};

let ExtensibleComponent = exports.ExtensibleComponent = (function (_Component) {
  function ExtensibleComponent() {
    let options = arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, ExtensibleComponent);

    _get(Object.getPrototypeOf(ExtensibleComponent.prototype), "constructor", this).call(this, options);

    this[_middlewareComponents] = [];
    this.subComponents.middlewareList = listComponent();
  }

  _inherits(ExtensibleComponent, _Component);

  _createClass(ExtensibleComponent, {
    addMiddleware: {
      value: function addMiddleware(middleware) {
        if (!middleware.isMiddlewareComponent) {
          throw new TypeError("middleware must be of type MiddlewareComponent");
        }

        this.middlewareList.push(middleware);

        return this;
      }
    },
    prependMiddleware: {
      value: function prependMiddleware(middleware) {
        if (!middleware.isMiddlewareComponent) {
          throw new TypeError("middleware must be of type MiddlewareComponent");
        }

        this.middlewareList.unshift(middleware);

        return this;
      }
    },
    middleware: {
      value: (function (_middleware) {
        var _middlewareWrapper = function middleware(_x) {
          return _middleware.apply(this, arguments);
        };

        _middlewareWrapper.toString = function () {
          return _middleware.toString();
        };

        return _middlewareWrapper;
      })(function (middleware) {
        return this.addMiddleware(middleware);
      })
    },
    toExtendMiddleware: {
      value: function toExtendMiddleware() {
        return combineMiddlewareComponents(this.middlewareComponents);
      }
    },
    middlewareList: {
      get: function () {
        return this.subComponents.middlewareList;
      }
    },
    middlewareComponents: {
      get: function () {
        return this.middlewareList.array;
      }
    }
  });

  return ExtensibleComponent;
})(Component);

let ExtensibleHandler = exports.ExtensibleHandler = (function (_ExtensibleComponent) {
  function ExtensibleHandler() {
    let options = arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, ExtensibleHandler);

    _get(Object.getPrototypeOf(ExtensibleHandler.prototype), "constructor", this).call(this, options);

    var _options$copyConfig = options.copyConfig;
    const copyConfig = _options$copyConfig === undefined ? true : _options$copyConfig;

    this._copyConfig = copyConfig;
  }

  _inherits(ExtensibleHandler, _ExtensibleComponent);

  _createClass(ExtensibleHandler, {
    toHandleableBuilder: {
      value: function toHandleableBuilder() {
        const copyConfig = this._copyConfig;

        const mainBuilder = this.toMainHandleableBuilder();
        const extendMiddleware = this.toExtendMiddleware();

        let builder = combineBuilderWithMiddleware(mainBuilder, extendMiddleware);

        if (copyConfig) builder = copyConfigBuilder(builder);

        return builder;
      }
    },
    toMainHandleableBuilder: {
      value: function toMainHandleableBuilder() {
        throw new Error("unimplemented");
      }
    },
    loadHandleable: {
      value: (function (_loadHandleable) {
        var _loadHandleableWrapper = function loadHandleable(_x2, _x3) {
          return _loadHandleable.apply(this, arguments);
        };

        _loadHandleableWrapper.toString = function () {
          return _loadHandleable.toString();
        };

        return _loadHandleableWrapper;
      })(function (config, options) {
        return loadHandleable(config, this.id, this.toHandleableBuilder(), options);
      })
    },
    loadHandler: {
      value: function loadHandler(config, options) {
        const loader = this.handlerLoader;
        return loader(config, this.id, this.toHandleableBuilder(), options);
      }
    },
    setLoader: {
      value: function setLoader(handlerLoader) {
        this._handlerLoader = handlerLoader;
        return this;
      }
    },
    handlerLoader: {
      get: function () {
        if (this._handlerLoader) return this._handlerLoader;

        return this.defaultLoader;
      }
    },
    defaultLoader: {
      get: function () {
        return loadHandleable;
      }
    },
    type: {
      get: function () {
        return "handler";
      }
    },
    isHandlerComponent: {
      get: function () {
        return true;
      }
    }
  });

  return ExtensibleHandler;
})(ExtensibleComponent);

let ExtensibleMiddleware = exports.ExtensibleMiddleware = (function (_ExtensibleComponent2) {
  function ExtensibleMiddleware() {
    _classCallCheck(this, ExtensibleMiddleware);

    if (_ExtensibleComponent2 != null) {
      _ExtensibleComponent2.apply(this, arguments);
    }
  }

  _inherits(ExtensibleMiddleware, _ExtensibleComponent2);

  _createClass(ExtensibleMiddleware, {
    toHandleableMiddleware: {
      value: function toHandleableMiddleware() {
        const mainMiddleware = this.toMainHandleableMiddleware();
        const extendMiddleware = this.toExtendMiddleware();

        return combineMiddlewares([mainMiddleware, extendMiddleware]);
      }
    },
    toMainHandleableMiddleware: {
      value: function toMainHandleableMiddleware() {
        throw new Error("unimplemented");
      }
    },
    isMiddlewareComponent: {
      get: function () {
        return true;
      }
    }
  });

  return ExtensibleMiddleware;
})(ExtensibleComponent);
