// quiver-component/out/lib/extend.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var loadHandleable = require("./util/loader").loadHandleable;

var HandleableBuilder = require("./handleable-builder").HandleableBuilder;

var HandleableMiddleware = require("./handleable-middleware").HandleableMiddleware;

let ExtendHandler = exports.ExtendHandler = (function (_HandleableBuilder) {
  function ExtendHandler(handlerComponent) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ExtendHandler);

    if (!handlerComponent.isHandlerComponent) {
      throw new Error("Extended component must be handler component");
    }

    options.safeWrapped = true;
    _get(Object.getPrototypeOf(ExtendHandler.prototype), "constructor", this).call(this, null, options);

    this.addSubComponent("extendHandler", handlerComponent);
  }

  _inherits(ExtendHandler, _HandleableBuilder);

  _createClass(ExtendHandler, {
    toMainHandleableBuilder: {
      value: function toMainHandleableBuilder() {
        const extendHandler = this.getSubComponent("extendHandler");
        const id = extendHandler.id;
        const builder = extendHandler.toHandleableBuilder();

        return function (config) {
          return loadHandleable(config, id, builder);
        };
      }
    }
  });

  return ExtendHandler;
})(HandleableBuilder);

let ExtendMiddleware = exports.ExtendMiddleware = (function (_HandleableMiddleware) {
  function ExtendMiddleware(middlewareComponent) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ExtendMiddleware);

    if (!middlewareComponent.isMiddlewareComponent) {
      throw new Error("Extended component must be middleware component");
    }

    options.safeWrapped = true;
    _get(Object.getPrototypeOf(ExtendMiddleware.prototype), "constructor", this).call(this, null, options);

    this.addSubComponent("extendMiddleware", middlewareComponent);
  }

  _inherits(ExtendMiddleware, _HandleableMiddleware);

  _createClass(ExtendMiddleware, {
    toMainHandleableBuilder: {
      value: function toMainHandleableBuilder() {
        return this.getSubComponent("extendMiddleware").toHandleableMiddleware();
      }
    }
  });

  return ExtendMiddleware;
})(HandleableMiddleware);

const extendHandler = function (handlerComponent) {
  return new ExtendHandler(handlerComponent);
};

exports.extendHandler = extendHandler;
const extendMiddleware = function (middlewareComponent) {
  return new ExtendMiddleware(middlewareComponent);
};
exports.extendMiddleware = extendMiddleware;
