// quiver-component/out/lib/filter.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var copy = require("quiver-object").copy;

var resolve = require("quiver-promise").resolve;

var streamToHttpHandler = require("quiver-http").streamToHttpHandler;

var _utilWrap = require("./util/wrap");

var safeHandler = _utilWrap.safeHandler;
var safeBuilder = _utilWrap.safeBuilder;

var HandleableMiddleware = require("./handleable-middleware").HandleableMiddleware;

const noCopy = function (config) {
  return config;
};

const filterToHandleableFilter = function (filter, handlerKey) {
  return function (config, handleable) {
    const handler = handleable[handlerKey];
    if (!handler) return resolve(handleable);

    return filter(config, handler).then(function (filteredHandler) {
      handleable[handlerKey] = filteredHandler;
      return handleable;
    });
  };
};

exports.filterToHandleableFilter = filterToHandleableFilter;
const streamToHandleableFilter = function (filter) {
  return filterToHandleableFilter(filter, "streamHandler");
};

exports.streamToHandleableFilter = streamToHandleableFilter;
const httpToHandleableFilter = function (filter) {
  return filterToHandleableFilter(filter, "httpHandler");
};

exports.httpToHandleableFilter = httpToHandleableFilter;
const filterToMiddleware = function (filter, copyConfig) {
  return function (config, builder) {
    return builder(copyConfig(config)).then(function (handler) {
      return filter(config, handler);
    });
  };
};

exports.filterToMiddleware = filterToMiddleware;

let HandleableFilter = exports.HandleableFilter = (function (_HandleableMiddleware) {
  function HandleableFilter(handleableFilter) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, HandleableFilter);

    handleableFilter = safeHandler(handleableFilter, options);

    _get(Object.getPrototypeOf(HandleableFilter.prototype), "constructor", this).call(this, null, options);

    var _options$copyConfig = options.copyConfig;
    const copyConfig = _options$copyConfig === undefined ? true : _options$copyConfig;

    this._copyConfig = copyConfig;

    this._handleableFilter = handleableFilter;
  }

  _inherits(HandleableFilter, _HandleableMiddleware);

  _createClass(HandleableFilter, {
    toMainHandleableMiddleware: {
      value: function toMainHandleableMiddleware() {
        const copyConfig = this._copyConfig ? copy : noCopy;
        const handleableFilter = this.toHandleableFilter();

        return filterToMiddleware(handleableFilter, copyConfig);
      }
    },
    toHandleableFilter: {
      value: function toHandleableFilter() {
        if (!this._handleableFilter) throw new Error("handleableFilter is not defined");

        return this._handleableFilter;
      }
    }
  });

  return HandleableFilter;
})(HandleableMiddleware);

let StreamFilter = exports.StreamFilter = (function (_HandleableFilter) {
  function StreamFilter(filter) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, StreamFilter);

    const streamFilter = safeBuilder(filter, options);
    _get(Object.getPrototypeOf(StreamFilter.prototype), "constructor", this).call(this, null, options);

    this._streamFilter = streamFilter;
  }

  _inherits(StreamFilter, _HandleableFilter);

  _createClass(StreamFilter, {
    toHandleableFilter: {
      value: function toHandleableFilter() {
        const streamFilter = this.toStreamFilter();

        return function (config, handleable) {
          const handler = handleable.streamHandler;
          if (!handler) return resolve(handleable);

          return streamFilter(config, handler).then(function (filteredHandler) {
            handleable.streamHandler = filteredHandler;
            return handleable;
          });
        };
      }
    },
    toStreamFilter: {
      value: function toStreamFilter() {
        const streamFilter = this._streamFilter;

        if (!streamFilter) throw new Error("streamFilter is not defined");

        return streamFilter;
      }
    },
    type: {
      get: function () {
        return "Stream Filter";
      }
    }
  });

  return StreamFilter;
})(HandleableFilter);

let HttpFilter = exports.HttpFilter = (function (_HandleableFilter2) {
  function HttpFilter(filter) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, HttpFilter);

    const httpFilter = safeBuilder(filter, options);
    _get(Object.getPrototypeOf(HttpFilter.prototype), "constructor", this).call(this, null, options);

    this._httpFilter = httpFilter;
  }

  _inherits(HttpFilter, _HandleableFilter2);

  _createClass(HttpFilter, {
    toHandleableFilter: {
      value: function toHandleableFilter() {
        const httpFilter = this.toHttpFilter();

        return function (config, handleable) {
          let httpHandler = handleable.httpHandler;
          if (!httpHandler) {
            const streamHandler = handleable.streamHandler;

            if (!streamHandler) return resolve(handleable);

            httpHandler = streamToHttpHandler(streamHandler);
            handleable = { httpHandler: httpHandler };
          }

          return httpFilter(config, httpHandler).then(function (filteredHandler) {
            handleable.httpHandler = filteredHandler;
            return handleable;
          });
        };
      }
    },
    toHttpFilter: {
      value: function toHttpFilter() {
        const httpFilter = this._httpFilter;

        if (!httpFilter) throw new Error("httpFilter is not defined");

        return httpFilter;
      }
    },
    type: {
      get: function () {
        return "Http Filter";
      }
    }
  });

  return HttpFilter;
})(HandleableFilter);

const handleableFilter = function (filter, options) {
  return new HandleableFilter(filter, options);
};

exports.handleableFilter = handleableFilter;
const streamFilter = function (filter, options) {
  return new StreamFilter(filter, options);
};

exports.streamFilter = streamFilter;
const httpFilter = function (filter, options) {
  return new HttpFilter(filter, options);
};
exports.httpFilter = httpFilter;
