// quiver-component/out/lib/handleable-builder.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var resolve = require("quiver-promise").resolve;

var safeHandler = require("./util/wrap").safeHandler;

var ExtensibleHandler = require("./extensible-component").ExtensibleHandler;

let HandleableBuilder = exports.HandleableBuilder = (function (_ExtensibleHandler) {
  function HandleableBuilder(handleableBuilder) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, HandleableBuilder);

    const mainHandleableBuilder = safeHandler(handleableBuilder, options);

    _get(Object.getPrototypeOf(HandleableBuilder.prototype), "constructor", this).call(this, options);

    this._mainHandleableBuilder = mainHandleableBuilder;
  }

  _inherits(HandleableBuilder, _ExtensibleHandler);

  _createClass(HandleableBuilder, {
    toMainHandleableBuilder: {
      value: function toMainHandleableBuilder() {
        return this._mainHandleableBuilder;
      }
    },
    type: {
      get: function () {
        return "handleable builder";
      }
    }
  });

  return HandleableBuilder;
})(ExtensibleHandler);

let Handleable = exports.Handleable = (function (_HandleableBuilder) {
  function Handleable(handleable) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, Handleable);

    options.safeWrapped = true;
    _get(Object.getPrototypeOf(Handleable.prototype), "constructor", this).call(this, null, options);

    this._handleable = handleable;
  }

  _inherits(Handleable, _HandleableBuilder);

  _createClass(Handleable, {
    toMainHandleableBuilder: {
      value: function toMainHandleableBuilder() {
        const handleable = this.toHandleable();

        return function (config) {
          return resolve(handleable);
        };
      }
    },
    toHandleable: {
      value: function toHandleable() {
        if (!this._handleable) throw new Error("handleable is not defined");

        return this._handleable;
      }
    },
    type: {
      get: function () {
        return "handleable";
      }
    }
  });

  return Handleable;
})(HandleableBuilder);

const handleableBuilder = function (builder, options) {
  return new HandleableBuilder(builder, options);
};

exports.handleableBuilder = handleableBuilder;
const handleable = function (handleable, options) {
  return new Handleable(handleable, options);
};
exports.handleable = handleable;
