// quiver-component/out/lib/http-handler.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var resolve = require("quiver-promise").resolve;

var loadHttpHandler = require("./util/loader").loadHttpHandler;

var _utilWrap = require("./util/wrap");

var safeBuilder = _utilWrap.safeBuilder;
var safeHandler = _utilWrap.safeHandler;

var HandleableBuilder = require("./handleable-builder").HandleableBuilder;

let HttpHandlerBuilder = exports.HttpHandlerBuilder = (function (_HandleableBuilder) {
  function HttpHandlerBuilder(httpHandlerBuilder) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, HttpHandlerBuilder);

    httpHandlerBuilder = safeBuilder(httpHandlerBuilder, options);

    _get(Object.getPrototypeOf(HttpHandlerBuilder.prototype), "constructor", this).call(this, null, options);

    this._httpHandlerBuilder = httpHandlerBuilder;
  }

  _inherits(HttpHandlerBuilder, _HandleableBuilder);

  _createClass(HttpHandlerBuilder, {
    toMainHandleableBuilder: {
      value: function toMainHandleableBuilder() {
        const builder = this.toHttpHandlerBuilder();

        return function (config) {
          return builder(config).then(function (httpHandler) {
            return { httpHandler: httpHandler };
          });
        };
      }
    },
    toHttpHandlerBuilder: {
      value: function toHttpHandlerBuilder() {
        if (!this._httpHandlerBuilder) throw new Error("httpHandlerBuilder is not defined");

        return this._httpHandlerBuilder;
      }
    },
    defaultLoader: {
      get: function () {
        return loadHttpHandler;
      }
    },
    type: {
      get: function () {
        return "Http Handler Builder";
      }
    }
  });

  return HttpHandlerBuilder;
})(HandleableBuilder);

let HttpHandler = exports.HttpHandler = (function (_HttpHandlerBuilder) {
  function HttpHandler(httpHandler) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, HttpHandler);

    httpHandler = safeHandler(httpHandler, options);

    _get(Object.getPrototypeOf(HttpHandler.prototype), "constructor", this).call(this, null, options);
    this._httpHandler = httpHandler;
  }

  _inherits(HttpHandler, _HttpHandlerBuilder);

  _createClass(HttpHandler, {
    toHttpHandlerBuilder: {
      value: function toHttpHandlerBuilder() {
        const handler = this.toHttpHandler();

        return function (config) {
          return resolve(handler);
        };
      }
    },
    toHttpHandler: {
      value: function toHttpHandler() {
        if (!this._httpHandler) throw new Error("httpHandler is not defined");

        return this._httpHandler;
      }
    },
    type: {
      get: function () {
        return "Http Handler";
      }
    }
  });

  return HttpHandler;
})(HttpHandlerBuilder);

const httpHandlerBuilder = function (builder, options) {
  return new HttpHandlerBuilder(builder, options);
};

exports.httpHandlerBuilder = httpHandlerBuilder;
const httpHandler = function (handler, options) {
  return new HttpHandler(handler, options);
};
exports.httpHandler = httpHandler;
