// quiver-component/out/lib/input-handler.js
"use strict";

var _interopRequire = function (obj) { return obj && obj.__esModule ? obj["default"] : obj; };

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var assertString = require("quiver-object").assertString;

var assign = _interopRequire(require("object.assign"));

var ConfigMiddleware = require("./simple-middleware").ConfigMiddleware;

var HandlerComponent = require("./component").HandlerComponent;

var ExtensibleComponent = require("./extensible-component").ExtensibleComponent;

let InputHandlerMiddleware = exports.InputHandlerMiddleware = (function (_ConfigMiddleware) {
  function InputHandlerMiddleware(handlerComponent, toConfig) {
    let options = arguments[2] === undefined ? {} : arguments[2];

    _classCallCheck(this, InputHandlerMiddleware);

    if (!handlerComponent.isHandlerComponent) {
      throw new TypeError("input handler must be of type HandlerComponent");
    }

    assertString(toConfig, "toConfig required to be string");

    var _options$loader = options.loader;
    const loader = _options$loader === undefined ? handlerComponent.handlerLoader : _options$loader;

    options.safeWrapped = true;

    _get(Object.getPrototypeOf(InputHandlerMiddleware.prototype), "constructor", this).call(this, null, options);

    this._handlerLoader = loader;
    this._toInputConfig = toConfig;

    this.subComponents.inputHandler = handlerComponent;
  }

  _inherits(InputHandlerMiddleware, _ConfigMiddleware);

  _createClass(InputHandlerMiddleware, {
    toConfigHandler: {
      value: function toConfigHandler() {
        const handlerComponent = this.inputHandlerComponent;

        const componentId = handlerComponent.id;
        const builder = handlerComponent.toHandleableBuilder();

        const loader = this._handlerLoader;
        const toConfig = this._toInputConfig;

        return function (config) {
          return loader(config, componentId, builder).then(function (handler) {
            config[toConfig] = handler;
          });
        };
      }
    },
    inputHandlerComponent: {
      get: function () {
        return this.subComponents.inputHandler;
      }
    },
    type: {
      get: function () {
        return "input handler middleware";
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(InputHandlerMiddleware.prototype), "toJson", this).call(this);
        json.inputHandler = this.inputHandlerComponent.toJson();
        return json;
      }
    }
  });

  return InputHandlerMiddleware;
})(ConfigMiddleware);

const InputHandlerMixin = {
  inputHandler: function inputHandler(handler, toConfig) {
    let options = arguments[2] === undefined ? {} : arguments[2];

    return this.middleware(new InputHandlerMiddleware(handler, toConfig, options));
  },
  inputHandlers: function inputHandlers(handlerMap) {
    for (let key in handlerMap) {
      const handler = handlerMap[key];
      this.inputHandler(handler, key);
    }
    return this;
  }
};

const mixinInputHandler = function (prototype) {
  return assign(prototype, InputHandlerMixin);
};

mixinInputHandler(ExtensibleComponent.prototype);

const inputHandlerMiddleware = function (handler, toConfig, options) {
  return new InputHandlerMiddleware(handler, toConfig, options);
};
exports.inputHandlerMiddleware = inputHandlerMiddleware;
