// quiver-component/out/lib/pipeline.js
"use strict";

var _toArray = function (arr) { return Array.isArray(arr) ? arr : Array.from(arr); };

var _toConsumableArray = function (arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } };

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var copy = require("quiver-object").copy;

var ExtensibleHandler = require("./extensible-component").ExtensibleHandler;

const combineStreamHandlers = function (handler1, handler2) {
  return function (args, streamable) {
    return handler1(copy(args), streamable).then(function (streamable) {
      return handler2(args, streamable);
    });
  };
};

const streamCombinator = {
  field: "streamHandler",
  combineHandlers: combineStreamHandlers
};

const defaultCombinators = [streamCombinator];

const combineHandleables = function (handleable1, handleable2, combinators) {
  const newHandleable = {};

  combinators.forEach(function (_ref) {
    let field = _ref.field;
    let combineHandlers = _ref.combineHandlers;

    const handler1 = handleable1[field];
    const handler2 = handleable2[field];

    if (!handler1 || !handler2) return;

    const newHandler = combineHandlers(handler1, handler2);
    newHandleable[field] = newHandler;
  });

  return newHandleable;
};

const pipelineHandleables = function (handleables, combinators) {
  if (handleables.length == 1) return handleables[0];

  var _handleables = _toArray(handleables);

  const handleable1 = _handleables[0];
  const handleable2 = _handleables[1];

  const restHandleables = _handleables.slice(2);

  const newHandleable = combineHandleables(handleable1, handleable2, combinators);

  return pipelineHandleables([newHandleable].concat(_toConsumableArray(restHandleables)), combinators);
};

const pipelineBuilder = function (builders, combinators) {
  return function (config) {
    return Promise.all(builders.map(function (builder) {
      return builder(config);
    })).then(function (handleables) {
      return pipelineHandleables(handleables, combinators);
    });
  };
};

let Pipeline = exports.Pipeline = (function (_ExtensibleHandler) {
  function Pipeline() {
    let options = arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, Pipeline);

    var _options$pipelineCombinators = options.pipelineCombinators;
    const pipelineCombinators = _options$pipelineCombinators === undefined ? defaultCombinators : _options$pipelineCombinators;

    _get(Object.getPrototypeOf(Pipeline.prototype), "constructor", this).call(this, options);

    this._pipelineCombinators = pipelineCombinators;
    this.subComponents.pipelineHandlers = [];
  }

  _inherits(Pipeline, _ExtensibleHandler);

  _createClass(Pipeline, {
    addPipe: {
      value: function addPipe(component) {
        if (!component.isHandlerComponent) {
          throw new TypeError("component must be of type HandlerComponent");
        }

        this.subComponents.pipelineHandlers.push(component);
        return this;
      }
    },
    pipe: {
      value: function pipe(component) {
        return this.addPipe(component);
      }
    },
    pipelineHandlers: {
      get: function () {
        return this.subComponents.pipelineHandlers;
      }
    },
    toMainHandleableBuilder: {
      value: function toMainHandleableBuilder() {
        const builders = this.pipelineHandlers.map(function (component) {
          return component.toHandleableBuilder();
        });

        if (builders.length == 0) throw new Error("Pipeline must contain at least one handler component");

        const combinators = this._pipelineCombinators;

        return pipelineBuilder(builders, combinators);
      }
    },
    type: {
      get: function () {
        return "pipeline";
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(Pipeline.prototype), "toJson", this).call(this);
        json.pipelines = this.pipelineHandlers.map(function (component) {
          return component.toJson();
        });

        return json;
      }
    }
  });

  return Pipeline;
})(ExtensibleHandler);

const pipeline = function (options) {
  return new Pipeline(options);
};
exports.pipeline = pipeline;
