// quiver-component/out/lib/simple-filter.js
"use strict";

var _interopRequire = function (obj) { return obj && obj.__esModule ? obj["default"] : obj; };

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var copy = require("quiver-object").copy;

var resolve = require("quiver-promise").resolve;

var assign = _interopRequire(require("object.assign"));

var StreamHandlerBuilder = require("./stream-handler").StreamHandlerBuilder;

var _filter = require("./filter");

var StreamFilter = _filter.StreamFilter;
var HttpFilter = _filter.HttpFilter;

var _utilWrap = require("./util/wrap");

var safeBuilder = _utilWrap.safeBuilder;
var safeHandler = _utilWrap.safeHandler;

const argsToStreamFilter = function (argsHandler) {
  return function (config, handler) {
    return resolve(function (args, inputStreamable) {
      return argsHandler(args).then(newArgs = function (args) {
        return handler(newArgs, inputStreamable);
      });
    });
  };
};

const errorToFilter = function (errorHandler) {
  return function (config, handler) {
    return resolve(function () {
      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return handler.apply(undefined, args)["catch"](function (err) {
        return errorHandler(err).then(function (result) {
          if (!result) throw err;
          return result;
        });
      });
    });
  };
};

const builderFilterConvert = function (builder, filterConvert) {
  return function (config, handler) {
    return builder(config).then(function (customHandler) {
      return filterConvert(customHandler)(config, handler);
    });
  };
};

const applyArgsFilter = function (argsHandler, handler) {
  return function (args, inputStreamable) {
    return argsHandler(args).then(function () {
      let newArgs = arguments[0] === undefined ? args : arguments[0];
      return handler(newArgs, inputStreamable);
    });
  };
};

const argsBuilderToFilter = function (argsBuilder) {
  return function (config, handler) {
    return argsBuilder(config).then(function (argsHandler) {
      return applyArgsFilter(argsHandler, handler);
    });
  };
};

let ArgsBuilderFilter = exports.ArgsBuilderFilter = (function (_StreamFilter) {
  function ArgsBuilderFilter(argsBuilder) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ArgsBuilderFilter);

    argsBuilder = safeBuilder(argsBuilder, options);

    if (options.copyConfig === undefined) options.copyConfig = true;

    _get(Object.getPrototypeOf(ArgsBuilderFilter.prototype), "constructor", this).call(this, null, options);

    this._argsBuilder = argsBuilder;
  }

  _inherits(ArgsBuilderFilter, _StreamFilter);

  _createClass(ArgsBuilderFilter, {
    toStreamFilter: {
      value: function toStreamFilter() {
        return argsBuilderToFilter(this.toArgsBuilder());
      }
    },
    toArgsBuilder: {
      value: function toArgsBuilder() {
        if (!this._argsBuilder) throw new Error("argsBuilder is not defined");

        return this._argsBuilder;
      }
    },
    type: {
      get: function () {
        return "args builder filter";
      }
    }
  });

  return ArgsBuilderFilter;
})(StreamFilter);

let ArgsFilter = exports.ArgsFilter = (function (_ArgsBuilderFilter) {
  function ArgsFilter(argsHandler) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ArgsFilter);

    argsHandler = safeHandler(argsHandler, options);
    options.copyConfig = false;

    _get(Object.getPrototypeOf(ArgsFilter.prototype), "constructor", this).call(this, null, options);
    this._argsHandler = argsHandler;
  }

  _inherits(ArgsFilter, _ArgsBuilderFilter);

  _createClass(ArgsFilter, {
    toArgsBuilder: {
      value: function toArgsBuilder() {
        const argsHandler = this.toArgsHandler();

        return function (config) {
          return resolve(argsHandler);
        };
      }
    },
    toArgsHandler: {
      value: function toArgsHandler() {
        if (!this._argsHandler) throw new Error("argsHandler is not defined");

        return this._argsHandler;
      }
    },
    type: {
      get: function () {
        return "args filter";
      }
    }
  });

  return ArgsFilter;
})(ArgsBuilderFilter);

let ErrorFilter = exports.ErrorFilter = (function (_StreamFilter2) {
  function ErrorFilter(errorHandler) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ErrorFilter);

    errorHandler = safeHandler(errorHandler, options);

    _get(Object.getPrototypeOf(ErrorFilter.prototype), "constructor", this).call(this, null, options);
    this._errorHandler = errorHandler;
  }

  _inherits(ErrorFilter, _StreamFilter2);

  _createClass(ErrorFilter, {
    toStreamFilter: {
      value: function toStreamFilter() {
        return errorToFilter(this.toErrorHandler());
      }
    },
    toErrorHandler: {
      value: function toErrorHandler() {
        if (!this._errorHandler) throw new Error("errorHandler is not defined");

        return this._errorHandler;
      }
    },
    type: {
      get: function () {
        return "error filter";
      }
    }
  });

  return ErrorFilter;
})(StreamFilter);

let ErrorBuilderFilter = exports.ErrorBuilderFilter = (function (_StreamFilter3) {
  function ErrorBuilderFilter(errorBuilder) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ErrorBuilderFilter);

    errorBuilder = safeBuilder(errorBuilder, options);

    _get(Object.getPrototypeOf(ErrorBuilderFilter.prototype), "constructor", this).call(this, null, options);
    this._errorBuilder = errorBuilder;
  }

  _inherits(ErrorBuilderFilter, _StreamFilter3);

  _createClass(ErrorBuilderFilter, {
    toStreamFilter: {
      value: function toStreamFilter() {
        return builderFilterConvert(this.toErrorBuilder(), errorToFilter);
      }
    },
    toErrorBuilder: {
      value: function toErrorBuilder() {
        if (!this._errorBuilder) throw new Error("errorBuilder is not defined");

        return this._errorBuilder;
      }
    },
    type: {
      get: function () {
        return "error builder filter";
      }
    }
  });

  return ErrorBuilderFilter;
})(StreamFilter);

const ArgsFilterMixin = {
  argsFilter: function argsFilter(argsHandler) {
    return this.middleware(new ArgsFilter(argsHandler));
  }
};

const mixinArgsFilter = function (prototype) {
  return assign(prototype, ArgsFilterMixin);
};

mixinArgsFilter(StreamHandlerBuilder.prototype);

const argsFilter = function (handler, options) {
  return new ArgsFilter(handler, options);
};

exports.argsFilter = argsFilter;
const argsBuilderFilter = function (builder, options) {
  return new ArgsBuilderFilter(builder, options);
};

exports.argsBuilderFilter = argsBuilderFilter;
const errorFilter = function (handler, options) {
  return new ErrorFilter(handler, options);
};

exports.errorFilter = errorFilter;
const errorBuilderFilter = function (builder, options) {
  return new ErrorBuilderFilter(builder, options);
};
exports.errorBuilderFilter = errorBuilderFilter;
