// quiver-component/out/lib/simple-handler.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var resolve = require("quiver-promise").resolve;

var _quiverSimpleHandler = require("quiver-simple-handler");

var simpleToStreamHandler = _quiverSimpleHandler.simpleToStreamHandler;
var streamToSimpleHandler = _quiverSimpleHandler.streamToSimpleHandler;
var validateSimpleTypes = _quiverSimpleHandler.validateSimpleTypes;

var simpleHandlerLoader = require("./util/loader").simpleHandlerLoader;

var _utilWrap = require("./util/wrap");

var safeBuilder = _utilWrap.safeBuilder;
var safeHandler = _utilWrap.safeHandler;

var _streamHandler = require("./stream-handler");

var StreamHandler = _streamHandler.StreamHandler;
var StreamHandlerBuilder = _streamHandler.StreamHandlerBuilder;

let SimpleHandlerBuilder = exports.SimpleHandlerBuilder = (function (_StreamHandlerBuilder) {
  function SimpleHandlerBuilder(simpleHandlerBuilder, inType, outType) {
    let options = arguments[3] === undefined ? {} : arguments[3];

    _classCallCheck(this, SimpleHandlerBuilder);

    const err = validateSimpleTypes([inType, outType]);
    if (err) throw err;

    simpleHandlerBuilder = safeBuilder(simpleHandlerBuilder, options);

    _get(Object.getPrototypeOf(SimpleHandlerBuilder.prototype), "constructor", this).call(this, null, options);

    this._inType = inType;
    this._outType = outType;
    this._simpleHandlerBuilder = simpleHandlerBuilder;
  }

  _inherits(SimpleHandlerBuilder, _StreamHandlerBuilder);

  _createClass(SimpleHandlerBuilder, {
    toStreamHandlerBuilder: {
      value: function toStreamHandlerBuilder() {
        var _ref = this;

        const inType = _ref.inType;
        const outType = _ref.outType;

        const simpleHandlerBuilder = this.toSimpleHandlerBuilder();

        return function (config) {
          return simpleHandlerBuilder(config).then(function (simpleHandler) {
            return simpleToStreamHandler(simpleHandler, inType, outType);
          });
        };
      }
    },
    toSimpleHandlerBuilder: {
      value: function toSimpleHandlerBuilder() {
        if (!this._simpleHandlerBuilder) throw new Error("simpleHandlerBuilder is not define");

        return this._simpleHandlerBuilder;
      }
    },
    defaultLoader: {
      get: function () {
        return simpleHandlerLoader(this.inType, this.outType);
      }
    },
    inType: {
      get: function () {
        return this._inType;
      }
    },
    outType: {
      get: function () {
        return this._outType;
      }
    },
    type: {
      get: function () {
        return "simple handler builder";
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(SimpleHandlerBuilder.prototype), "toJson", this).call(this);

        json.inType = this.inType;
        json.outType = this.outType;

        return json;
      }
    }
  });

  return SimpleHandlerBuilder;
})(StreamHandlerBuilder);

let SimpleHandler = exports.SimpleHandler = (function (_SimpleHandlerBuilder) {
  function SimpleHandler(simpleHandler, inType, outType) {
    let options = arguments[3] === undefined ? {} : arguments[3];

    _classCallCheck(this, SimpleHandler);

    simpleHandler = safeHandler(simpleHandler, options);

    _get(Object.getPrototypeOf(SimpleHandler.prototype), "constructor", this).call(this, null, inType, outType, options);
    this._simpleHandler = simpleHandler;
  }

  _inherits(SimpleHandler, _SimpleHandlerBuilder);

  _createClass(SimpleHandler, {
    toSimpleHandlerBuilder: {
      value: function toSimpleHandlerBuilder() {
        const simpleHandler = this.toSimpleHandler();

        return function (config) {
          return resolve(simpleHandler);
        };
      }
    },
    toSimpleHandler: {
      value: function toSimpleHandler() {
        if (!this._simpleHandler) throw new Error("simpleHandler is not defined");

        return this._simpleHandler;
      }
    },
    type: {
      get: function () {
        return "simple handler";
      }
    }
  });

  return SimpleHandler;
})(SimpleHandlerBuilder);

const simpleHandlerBuilder = function (builder, inType, outType, options) {
  return new SimpleHandlerBuilder(builder, inType, outType, options);
};

exports.simpleHandlerBuilder = simpleHandlerBuilder;
const simpleHandler = function (handler, inType, outType, options) {
  return new SimpleHandler(handler, inType, outType, options);
};
exports.simpleHandler = simpleHandler;
