// quiver-component/out/lib/simple-middleware.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var copy = require("quiver-object").copy;

var resolve = require("quiver-promise").resolve;

var safeHandler = require("./util/wrap").safeHandler;

var HandleableMiddleware = require("./handleable-middleware").HandleableMiddleware;

var ExtensibleComponent = require("./extensible-component").ExtensibleComponent;

const configHandlerToMiddleware = function (configHandler) {
  return function (config, builder) {
    return configHandler(config).then(function () {
      let newConfig = arguments[0] === undefined ? config : arguments[0];
      return builder(newConfig);
    });
  };
};

let ConfigMiddleware = exports.ConfigMiddleware = (function (_HandleableMiddleware) {
  function ConfigMiddleware(configHandler) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ConfigMiddleware);

    configHandler = safeHandler(configHandler, options);

    _get(Object.getPrototypeOf(ConfigMiddleware.prototype), "constructor", this).call(this, null, options);
    this._configHandler = configHandler;
  }

  _inherits(ConfigMiddleware, _HandleableMiddleware);

  _createClass(ConfigMiddleware, {
    toMainHandleableMiddleware: {
      value: function toMainHandleableMiddleware() {
        const configHandler = this.toConfigHandler();

        return configHandlerToMiddleware(configHandler);
      }
    },
    toConfigHandler: {
      value: function toConfigHandler() {
        if (!this._configHandler) throw new Error("configHandler is not defined");

        return this._configHandler;
      }
    },
    type: {
      get: function () {
        return "config middleware";
      }
    }
  });

  return ConfigMiddleware;
})(HandleableMiddleware);

let ConfigOverrideMiddleware = exports.ConfigOverrideMiddleware = (function (_ConfigMiddleware) {
  function ConfigOverrideMiddleware(overrideConfig) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ConfigOverrideMiddleware);

    options.safeWrapped = true;

    _get(Object.getPrototypeOf(ConfigOverrideMiddleware.prototype), "constructor", this).call(this, null, options);
    this._overrideConfig = overrideConfig;
  }

  _inherits(ConfigOverrideMiddleware, _ConfigMiddleware);

  _createClass(ConfigOverrideMiddleware, {
    toConfigHandler: {
      value: function toConfigHandler() {
        const overrideConfig = this.overrideConfig;

        return function (config) {
          for (let key in overrideConfig) {
            config[key] = overrideConfig[key];
          }

          return resolve(config);
        };
      }
    },
    overrideConfig: {
      get: function () {
        return this._overrideConfig;
      }
    },
    type: {
      get: function () {
        return "config override middleware";
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(ConfigOverrideMiddleware.prototype), "toJson", this).call(this);

        json.overrideConfig = this.overrideConfig;
        return json;
      }
    }
  });

  return ConfigOverrideMiddleware;
})(ConfigMiddleware);

let ConfigAliasMiddleware = exports.ConfigAliasMiddleware = (function (_ConfigMiddleware2) {
  function ConfigAliasMiddleware(aliasConfig) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ConfigAliasMiddleware);

    options.safeWrapped = true;

    _get(Object.getPrototypeOf(ConfigAliasMiddleware.prototype), "constructor", this).call(this, null, options);
    this._aliasConfig = aliasConfig;
  }

  _inherits(ConfigAliasMiddleware, _ConfigMiddleware2);

  _createClass(ConfigAliasMiddleware, {
    toConfigHandler: {
      value: function toConfigHandler() {
        const aliasConfig = this._aliasConfig;

        return function (config) {
          for (let key in aliasConfig) {
            const aliasKey = aliasConfig[key];
            config[key] = config[aliasKey];
          }

          return resolve(config);
        };
      }
    },
    aliasConfig: {
      get: function () {
        return copy(this._aliasConfig);
      }
    },
    type: {
      get: function () {
        return "config alias middleware";
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(ConfigAliasMiddleware.prototype), "toJson", this).call(this);

        json.aliasConfig = this.aliasConfig;
        return json;
      }
    }
  });

  return ConfigAliasMiddleware;
})(ConfigMiddleware);

const mixinConfigOverride = function (prototype) {
  prototype.configOverride = function (config) {
    return this.addMiddleware(new ConfigOverrideMiddleware(config));
  };
};

const mixinConfigAlias = function (prototype) {
  prototype.configAlias = function (config) {
    return this.addMiddleware(new ConfigAliasMiddleware(config));
  };
};

mixinConfigOverride(ExtensibleComponent.prototype);
mixinConfigAlias(ExtensibleComponent.prototype);

const configMiddleware = function (handler) {
  return new ConfigMiddleware(handler);
};

exports.configMiddleware = configMiddleware;
const configOverrideMiddleware = function (config) {
  return new ConfigOverrideMiddleware(config);
};

exports.configOverrideMiddleware = configOverrideMiddleware;
const configAliasMiddleware = function (config) {
  return new ConfigAliasMiddleware(config);
};
exports.configAliasMiddleware = configAliasMiddleware;
