// quiver-component/out/lib/stream-handler.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var resolve = require("quiver-promise").resolve;

var loadStreamHandler = require("./util/loader").loadStreamHandler;

var _utilWrap = require("./util/wrap");

var safeBuilder = _utilWrap.safeBuilder;
var safeHandler = _utilWrap.safeHandler;

var HandleableBuilder = require("./handleable-builder").HandleableBuilder;

let StreamHandlerBuilder = exports.StreamHandlerBuilder = (function (_HandleableBuilder) {
  function StreamHandlerBuilder(streamHandlerBuilder) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, StreamHandlerBuilder);

    streamHandlerBuilder = safeBuilder(streamHandlerBuilder, options);

    _get(Object.getPrototypeOf(StreamHandlerBuilder.prototype), "constructor", this).call(this, null, options);
    this._streamHandlerBuilder = streamHandlerBuilder;
  }

  _inherits(StreamHandlerBuilder, _HandleableBuilder);

  _createClass(StreamHandlerBuilder, {
    toMainHandleableBuilder: {
      value: function toMainHandleableBuilder() {
        const builder = this.toStreamHandlerBuilder();

        return function (config) {
          return builder(config).then(function (streamHandler) {
            return { streamHandler: streamHandler };
          });
        };
      }
    },
    toStreamHandlerBuilder: {
      value: function toStreamHandlerBuilder() {
        if (!this._streamHandlerBuilder) throw new Error("streamHandlerBuilder is not defined");

        return this._streamHandlerBuilder;
      }
    },
    defaultLoader: {
      get: function () {
        return loadStreamHandler;
      }
    },
    type: {
      get: function () {
        return "stream handler builder";
      }
    }
  });

  return StreamHandlerBuilder;
})(HandleableBuilder);

let StreamHandler = exports.StreamHandler = (function (_StreamHandlerBuilder) {
  function StreamHandler(streamHandler) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, StreamHandler);

    streamHandler = safeHandler(streamHandler, options);

    _get(Object.getPrototypeOf(StreamHandler.prototype), "constructor", this).call(this, null, options);
    this._streamHandler = streamHandler;
  }

  _inherits(StreamHandler, _StreamHandlerBuilder);

  _createClass(StreamHandler, {
    toStreamHandlerBuilder: {
      value: function toStreamHandlerBuilder() {
        const handler = this.toStreamHandler();

        return function (config) {
          return resolve(handler);
        };
      }
    },
    toStreamHandler: {
      value: function toStreamHandler() {
        if (!this._streamHandler) throw new Error("streamHandler is not defined");

        return this._streamHandler;
      }
    },
    type: {
      get: function () {
        return "stream handler";
      }
    }
  });

  return StreamHandler;
})(StreamHandlerBuilder);

const streamHandlerBuilder = function (builder, options) {
  return new StreamHandlerBuilder(builder, options);
};

exports.streamHandlerBuilder = streamHandlerBuilder;
const streamHandler = function (handler, options) {
  return new StreamHandler(handler, options);
};
exports.streamHandler = streamHandler;
