// quiver-component/out/lib/transform-filter.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var resolve = require("quiver-promise").resolve;

var copy = require("quiver-object").copy;

var StreamFilter = require("./filter").StreamFilter;

var HandlerComponent = require("./component").HandlerComponent;

var loadStreamHandler = require("./util/loader").loadStreamHandler;

const validModes = {
  "in": true,
  out: true,
  inout: true
};

const echoHandler = function (args, streamable) {
  return resolve(streamable);
};

const wrapHandler = function (handler) {
  return function (args) {
    for (var _len = arguments.length, restArgs = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      restArgs[_key - 1] = arguments[_key];
    }

    return handler.apply(undefined, [copy(args)].concat(restArgs));
  };
};

const inTransformHandler = function (handler, mode) {
  return mode != "out" ? wrapHandler(handler) : echoHandler;
};

const wrapMainHandler = function (handler, mode) {
  return mode == "in" ? handler : wrapHandler(handler);
};

const outTransformHandler = function (handler, mode) {
  return mode != "in" ? handler : echoHandler;
};

let TransformFilter = exports.TransformFilter = (function (_StreamFilter) {
  function TransformFilter(handlerComponent, transformMode) {
    let options = arguments[2] === undefined ? {} : arguments[2];

    _classCallCheck(this, TransformFilter);

    if (!handlerComponent.isHandlerComponent) throw new TypeError("input handler component must be of type HandlerComponent");

    if (!validModes[transformMode]) throw new TypeError("invalid transform mode provided in options");

    options.safeWrapped = true;

    _get(Object.getPrototypeOf(TransformFilter.prototype), "constructor", this).call(this, null, options);

    this._transformMode = transformMode;
    this.subComponents.transformComponent = handlerComponent;
  }

  _inherits(TransformFilter, _StreamFilter);

  _createClass(TransformFilter, {
    toStreamFilter: {
      value: function toStreamFilter() {
        const transformComponent = this.transformComponent;

        const componentId = transformComponent.id;
        const builder = transformComponent.toHandleableBuilder();

        const transformMode = this.transformMode;

        return function (config, handler) {
          return loadStreamHandler(config, componentId, builder).then(function (transformHandler) {
            const transformIn = inTransformHandler(transformHandler, transformMode);

            const mainHandler = wrapMainHandler(handler, transformMode);

            const transformOut = outTransformHandler(transformHandler, transformMode);

            return function (args, streamable) {
              return transformIn(args, streamable).then(function (transformedIn) {
                return mainHandler(args, transformedIn).then(function (resultStreamable) {
                  return transformOut(args, resultStreamable);
                });
              });
            };
          });
        };
      }
    },
    transformComponent: {
      get: function () {
        return this.subComponents.transformComponent;
      }
    },
    transformMode: {
      get: function () {
        return this._transformMode;
      }
    },
    type: {
      get: function () {
        return "transform filter";
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(TransformFilter.prototype), "toJson", this).call(this);

        json.transformMode = this.transformMode;
        json.transformHandler = this.transformComponent.toJson();

        return json;
      }
    }
  });

  return TransformFilter;
})(StreamFilter);

const transformFilter = function (handler, mode, options) {
  return new TransformFilter(handler, mode, options);
};
exports.transformFilter = transformFilter;
