// quiver-simple-handler/out/lib/simple-handler.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _quiverStreamUtil = require('quiver-stream-util');

var _quiverError = require('quiver-error');

var _quiverPromise = require('quiver-promise');

const convertHandler = function convertHandler(handler, inConvert, outConvert) {
  return function (args, input) {
    return (0, _quiverPromise.resolve)(inConvert(input)).then(function (input) {
      return handler(args, input).then(function (result) {
        return outConvert(result);
      });
    });
  };
};

const streamableToVoid = function streamableToVoid(streamable) {
  if (streamable.reusable) return (0, _quiverPromise.resolve)();

  return streamable.toStream().then(function (readStream) {
    return readStream.closeRead();
  });
};

const voidToStreamable = function voidToStreamable() {
  return (0, _quiverPromise.resolve)((0, _quiverStreamUtil.emptyStreamable)());
};

const streamableToStream = function streamableToStream(streamable) {
  return streamable.toStream();
};

const streamableToStreamable = function streamableToStreamable(streamable) {
  return (0, _quiverPromise.resolve)(streamable);
};

const htmlToStreamable = function htmlToStreamable(text) {
  return (0, _quiverStreamUtil.textToStreamable)(text, 'text/html');
};

const streamToSimpleTable = {
  'void': streamableToVoid,
  'text': _quiverStreamUtil.streamableToText,
  'string': _quiverStreamUtil.streamableToText,
  'html': _quiverStreamUtil.streamableToText,
  'json': _quiverStreamUtil.streamableToJson,
  'stream': streamableToStream,
  'streamable': streamableToStreamable
};

const simpleToStreamTable = {
  'void': voidToStreamable,
  'text': _quiverStreamUtil.textToStreamable,
  'string': _quiverStreamUtil.textToStreamable,
  'html': htmlToStreamable,
  'json': _quiverStreamUtil.jsonToStreamable,
  'stream': _quiverStreamUtil.streamToStreamable,
  'streamable': streamableToStreamable
};

const createConverter = function createConverter(inTable, outTable) {
  return function (handler, inType, outType) {
    const inConvert = inTable[inType];
    if (!inConvert) throw new Error('invalid simple type ' + inType);

    const outConvert = outTable[outType];
    if (!outConvert) throw new Error('invalid simple type ' + outType);

    return convertHandler((0, _quiverPromise.safePromised)(handler), inConvert, outConvert);
  };
};

const simpleToStreamHandler = createConverter(streamToSimpleTable, simpleToStreamTable);

exports.simpleToStreamHandler = simpleToStreamHandler;
const streamToSimpleHandler = createConverter(simpleToStreamTable, streamToSimpleTable);

exports.streamToSimpleHandler = streamToSimpleHandler;
const validateSimpleTypes = function validateSimpleTypes(types) {
  for (let type of types) {
    if (!type) {
      return new Error('simple type is not defined');
    }

    if (typeof type != 'string') {
      return new Error('simple type must be of type string');
    }

    if (!streamToSimpleTable[type]) {
      return new Error('invalid simple type ' + type);
    }
  }
};
exports.validateSimpleTypes = validateSimpleTypes;
