// quiver-stream-util/out/lib/buffer.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _quiverPromise = require('quiver-promise');

var _quiverStreamChannel = require('quiver-stream-channel');

var _buffers = require('./buffers');

const nodeBuffers = function nodeBuffers(buffers) {
  return buffers.map(function (buffer) {
    return buffer instanceof Buffer ? buffer : new Buffer(buffer);
  });
};

const buffersToBuffer = function buffersToBuffer(buffers) {
  return Buffer.concat(nodeBuffers(buffers));
};

const streamToBuffer = function streamToBuffer(readStream) {
  return (0, _buffers.streamToBuffers)(readStream).then(buffersToBuffer);
};

exports.streamToBuffer = streamToBuffer;
const streamableToBuffer = function streamableToBuffer(streamable) {
  if (streamable.toBuffer) return (0, _quiverPromise.resolve)(streamable.toBuffer());

  return (0, _buffers.streamableToBuffers)(streamable).then(buffersToBuffer).then(function (buffer) {
    if (streamable.reusable && !streamable.toBuffer) {
      streamable.toBuffer = function () {
        return (0, _quiverPromise.resolve)(buffer);
      };
    }

    return buffer;
  });
};

exports.streamableToBuffer = streamableToBuffer;
const bufferToStream = function bufferToStream(buffer) {
  if (!Buffer.isBuffer(buffer)) buffer = new Buffer(buffer);

  var _createChannel = (0, _quiverStreamChannel.createChannel)();

  const readStream = _createChannel.readStream;
  const writeStream = _createChannel.writeStream;

  writeStream.write(buffer);
  writeStream.closeWrite(null);

  return readStream;
};

exports.bufferToStream = bufferToStream;
const toBufferToStreamable = function toBufferToStreamable(_toBuffer) {
  return Object.defineProperties({
    reusable: true,
    toBuffer: function toBuffer() {
      return (0, _quiverPromise.resolve)(_toBuffer());
    },
    toBuffers: function toBuffers() {
      return (0, _quiverPromise.resolve)([_toBuffer()]);
    },
    toStream: function toStream() {
      return (0, _quiverPromise.resolve)(bufferToStream(_toBuffer()));
    }
  }, {
    contentLength: {
      get: function () {
        return _toBuffer().length;
      },
      configurable: true,
      enumerable: true
    }
  });
};

exports.toBufferToStreamable = toBufferToStreamable;
const bufferToStreamable = function bufferToStreamable(buffer) {
  if (!Buffer.isBuffer(buffer)) buffer = new Buffer(buffer);

  return toBufferToStreamable(function () {
    return buffer;
  });
};
exports.bufferToStreamable = bufferToStreamable;
