// quiver-stream-util/out/lib/buffers.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _quiverPromise = require('quiver-promise');

var _quiverStreamChannel = require('quiver-stream-channel');

const streamToBuffers = function streamToBuffers(readStream) {
  const buffers = [];

  const doPipe = function doPipe(callback) {
    return readStream.read().then(function (_ref) {
      let closed = _ref.closed;
      let data = _ref.data;

      if (closed) return callback(null, buffers);

      buffers.push(data);
      doPipe(callback);
    }, callback);
  };

  return (0, _quiverPromise.promisify)(doPipe)();
};

exports.streamToBuffers = streamToBuffers;
const streamableToBuffers = function streamableToBuffers(streamable) {
  if (streamable.toBuffers) return (0, _quiverPromise.resolve)(streamable.toBuffers());

  return streamable.toStream().then(function (readStream) {
    return streamToBuffers(readStream).then(function (buffers) {
      if (streamable.reusable && !streamable.toBuffers) {
        streamable.toBuffers = function () {
          return (0, _quiverPromise.resolve)(buffers.slice());
        };
        return buffers.slice();
      }

      return buffers;
    });
  });
};

exports.streamableToBuffers = streamableToBuffers;
const buffersToStream = function buffersToStream(buffers) {
  var _createChannel = (0, _quiverStreamChannel.createChannel)();

  const readStream = _createChannel.readStream;
  const writeStream = _createChannel.writeStream;

  buffers.forEach(function (buffer) {
    return writeStream.write(buffer);
  });
  writeStream.closeWrite();

  return readStream;
};

exports.buffersToStream = buffersToStream;
const buffersToStreamable = function buffersToStreamable(buffers) {
  return {
    reusable: true,
    toBuffers: function toBuffers() {
      return (0, _quiverPromise.resolve)(buffers.slice());
    },
    toStream: function toStream() {
      return (0, _quiverPromise.resolve)(buffersToStream(buffers));
    }
  };
};
exports.buffersToStreamable = buffersToStreamable;
