// quiver-stream-util/out/lib/empty.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _quiverPromise = require('quiver-promise');

const emptyReadStream = function emptyReadStream() {
  return {
    read: function read() {
      return (0, _quiverPromise.resolve)({ closed: true });
    },
    closeRead: function closeRead(err) {},
    isClosed: function isClosed() {
      return {};
    }
  };
};

exports.emptyReadStream = emptyReadStream;
const emptyWriteStream = function emptyWriteStream(closeErr) {
  return {
    prepareWrite: function prepareWrite() {
      return (0, _quiverPromise.resolve)({ closed: true });
    },
    write: function write(data) {},
    closeWrite: function closeWrite(err) {},
    isClosed: function isClosed() {
      return {};
    }
  };
};

exports.emptyWriteStream = emptyWriteStream;
const emptyStreamable = function emptyStreamable(closeErr) {
  return {
    reusable: true,
    contentLength: 0,
    toStream: function toStream() {
      return (0, _quiverPromise.resolve)(emptyReadStream(closeErr));
    },
    toText: function toText() {
      return (0, _quiverPromise.resolve)('');
    },
    toBuffer: function toBuffer() {
      return (0, _quiverPromise.resolve)(new Buffer(0));
    }
  };
};
exports.emptyStreamable = emptyStreamable;
