// quiver-stream-util/out/lib/node-stream.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _quiverError = require('quiver-error');

var _quiverPromise = require('quiver-promise');

var _quiverStreamChannel = require('quiver-stream-channel');

const noop = function noop() {};

const nodeReadToStreamable = function nodeReadToStreamable(nodeRead) {
  let opened = false;

  const toStream = function toStream() {
    if (opened) return (0, _quiverPromise.reject)((0, _quiverError.error)(500, 'streamable can only be opened once'));

    opened = true;
    return (0, _quiverPromise.resolve)(nodeToQuiverReadStream(nodeRead));
  };

  const toNodeStream = function toNodeStream() {
    if (opened) return (0, _quiverPromise.reject)((0, _quiverError.error)(500, 'streamable can only be opened once'));

    opened = true;
    return (0, _quiverPromise.resolve)(nodeRead);
  };

  return {
    reusable: false,
    toStream: toStream,
    toNodeStream: toNodeStream
  };
};

exports.nodeReadToStreamable = nodeReadToStreamable;
const nodeToQuiverReadStream = function nodeToQuiverReadStream(nodeRead) {
  var _createChannel = (0, _quiverStreamChannel.createChannel)();

  const readStream = _createChannel.readStream;
  const writeStream = _createChannel.writeStream;

  let ended = false;

  nodeRead.on('end', function () {
    if (ended) return;
    ended = true;

    writeStream.closeWrite();
  });

  nodeRead.on('error', function (err) {
    if (ended) return;
    ended = true;

    writeStream.closeWrite(err);
  });

  const doRead = function doRead(callback) {
    if (ended) return;

    const data = nodeRead.read();
    if (data) return callback(data);

    nodeRead.once('readable', function () {
      return doRead(callback);
    });
  };

  const doPipe = function doPipe() {
    return writeStream.prepareWrite().then(function (_ref) {
      let closed = _ref.closed;

      // force the node read stream into flowing mode
      // because there is no way to cancel a node read stream?!
      if (closed) return nodeRead.resume();

      doRead(function (data) {
        writeStream.write(data);
        doPipe();
      });
    });
  };

  doPipe();

  return readStream;
};

exports.nodeToQuiverReadStream = nodeToQuiverReadStream;
const nodeToQuiverWriteStream = function nodeToQuiverWriteStream(nodeWrite) {
  var _createChannel2 = (0, _quiverStreamChannel.createChannel)();

  const readStream = _createChannel2.readStream;
  const writeStream = _createChannel2.writeStream;

  nodeWrite.on('error', function (err) {
    return readStream.closeRead(err);
  });

  const doPipe = function doPipe() {
    return readStream.read().then(function (_ref2) {
      let closed = _ref2.closed;
      let data = _ref2.data;

      if (closed) return nodeWrite.end();

      const ready = nodeWrite.write(data);
      if (ready) return doPipe();

      nodeWrite.once('drain', doPipe);
    }, function (err) {
      return nodeWrite.end();
    });
  };

  doPipe();

  return writeStream;
};
exports.nodeToQuiverWriteStream = nodeToQuiverWriteStream;
