// quiver-stream-util/out/lib/streamable.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _quiverError = require('quiver-error');

var _quiverPromise = require('quiver-promise');

var _quiverStreamChannel = require('quiver-stream-channel');

var _buffers = require('./buffers');

const closeStreamable = function closeStreamable(streamable) {
  if (streamable.reusable) return (0, _quiverPromise.resolve)();

  return streamable.toStream().then(function (readStream) {
    return readStream.closeRead();
  });
};

exports.closeStreamable = closeStreamable;
const streamToStreamable = function streamToStreamable(readStream) {
  let opened = false;

  return {
    reusable: false,
    toStream: function toStream() {
      if (opened) return reject((0, _quiverError.error)(500, 'streamable can only be opened once'));

      opened = true;
      return (0, _quiverPromise.resolve)(readStream);
    }
  };
};

exports.streamToStreamable = streamToStreamable;
const reuseStream = function reuseStream(readStream) {
  const streamable = {
    reusable: true
  };

  const buffers = [];
  const pendingWrites = [];

  streamable.toStream = function () {
    var _createChannel = (0, _quiverStreamChannel.createChannel)();

    const readStream = _createChannel.readStream;
    const writeStream = _createChannel.writeStream;

    buffers.forEach(function (buffer) {
      return writeStream.write(buffer);
    });
    pendingWrites.push(writeStream);
    return (0, _quiverPromise.resolve)(readStream);
  };

  const doPipe = function doPipe() {
    readStream.read().then(function (_ref) {
      let closed = _ref.closed;
      let data = _ref.data;

      if (closed) {
        pendingWrites.forEach(function (writeStream) {
          return writeStream.closeWrite();
        });

        const allBuffers = buffers;
        streamable.toStream = function () {
          return (0, _quiverPromise.resolve)((0, _buffers.buffersToStream)(allBuffers));
        };
      } else {
        buffers.push(data);

        pendingWrites.forEach(function (writeStream) {
          return writeStream.write(data);
        });

        doPipe();
      }
    }, function (err) {
      pendingWrites.forEach(function (writeStream) {
        return writeStream.closeWrite(err);
      });

      streamable.toStream = function () {
        return reject(err);
      };
    });
  };

  doPipe();

  return streamable;
};

exports.reuseStream = reuseStream;
const reuseStreamable = function reuseStreamable(streamable) {
  if (streamable.reusable) return (0, _quiverPromise.resolve)(streamable);

  return streamable.toStream().then(reuseStream);
};

exports.reuseStreamable = reuseStreamable;
const unreuseStreamable = function unreuseStreamable(streamable) {
  if (!streamable.reusable) return streamable;

  const oldToStream = streamable.toStream;

  streamable.reusable = false;

  let opened = false;
  streamable.toStream = function () {
    if (opened) return rreject((0, _quiverError.error)(500, 'streamable can only be opened once'));

    opened = true;
    return oldToStream();
  };

  return streamable;
};
exports.unreuseStreamable = unreuseStreamable;
