// quiver-stream-util/out/lib/stream-util.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _buffers = require('./buffers');

Object.defineProperty(exports, 'streamToBuffers', {
  enumerable: true,
  get: function get() {
    return _buffers.streamToBuffers;
  }
});
Object.defineProperty(exports, 'streamableToBuffers', {
  enumerable: true,
  get: function get() {
    return _buffers.streamableToBuffers;
  }
});
Object.defineProperty(exports, 'buffersToStream', {
  enumerable: true,
  get: function get() {
    return _buffers.buffersToStream;
  }
});
Object.defineProperty(exports, 'buffersToStreamable', {
  enumerable: true,
  get: function get() {
    return _buffers.buffersToStreamable;
  }
});

var _buffer = require('./buffer');

Object.defineProperty(exports, 'streamToBuffer', {
  enumerable: true,
  get: function get() {
    return _buffer.streamToBuffer;
  }
});
Object.defineProperty(exports, 'streamableToBuffer', {
  enumerable: true,
  get: function get() {
    return _buffer.streamableToBuffer;
  }
});
Object.defineProperty(exports, 'bufferToStream', {
  enumerable: true,
  get: function get() {
    return _buffer.bufferToStream;
  }
});
Object.defineProperty(exports, 'bufferToStreamable', {
  enumerable: true,
  get: function get() {
    return _buffer.bufferToStreamable;
  }
});

var _text = require('./text');

Object.defineProperty(exports, 'streamToText', {
  enumerable: true,
  get: function get() {
    return _text.streamToText;
  }
});
Object.defineProperty(exports, 'streamableToText', {
  enumerable: true,
  get: function get() {
    return _text.streamableToText;
  }
});
Object.defineProperty(exports, 'textToStream', {
  enumerable: true,
  get: function get() {
    return _text.textToStream;
  }
});
Object.defineProperty(exports, 'textToStreamable', {
  enumerable: true,
  get: function get() {
    return _text.textToStreamable;
  }
});

var _json = require('./json');

Object.defineProperty(exports, 'streamToJson', {
  enumerable: true,
  get: function get() {
    return _json.streamToJson;
  }
});
Object.defineProperty(exports, 'streamableToJson', {
  enumerable: true,
  get: function get() {
    return _json.streamableToJson;
  }
});
Object.defineProperty(exports, 'jsonToStream', {
  enumerable: true,
  get: function get() {
    return _json.jsonToStream;
  }
});
Object.defineProperty(exports, 'jsonToStreamable', {
  enumerable: true,
  get: function get() {
    return _json.jsonToStreamable;
  }
});

var _streamable = require('./streamable');

Object.defineProperty(exports, 'closeStreamable', {
  enumerable: true,
  get: function get() {
    return _streamable.closeStreamable;
  }
});
Object.defineProperty(exports, 'streamToStreamable', {
  enumerable: true,
  get: function get() {
    return _streamable.streamToStreamable;
  }
});
Object.defineProperty(exports, 'reuseStream', {
  enumerable: true,
  get: function get() {
    return _streamable.reuseStream;
  }
});
Object.defineProperty(exports, 'reuseStreamable', {
  enumerable: true,
  get: function get() {
    return _streamable.reuseStreamable;
  }
});
Object.defineProperty(exports, 'unreuseStreamable', {
  enumerable: true,
  get: function get() {
    return _streamable.unreuseStreamable;
  }
});

var _pushback = require('./pushback');

Object.defineProperty(exports, 'pushbackStream', {
  enumerable: true,
  get: function get() {
    return _pushback.pushbackStream;
  }
});

var _nodeStream = require('./node-stream');

Object.defineProperty(exports, 'nodeToQuiverReadStream', {
  enumerable: true,
  get: function get() {
    return _nodeStream.nodeToQuiverReadStream;
  }
});
Object.defineProperty(exports, 'nodeToQuiverWriteStream', {
  enumerable: true,
  get: function get() {
    return _nodeStream.nodeToQuiverWriteStream;
  }
});
Object.defineProperty(exports, 'nodeReadToStreamable', {
  enumerable: true,
  get: function get() {
    return _nodeStream.nodeReadToStreamable;
  }
});

var _empty = require('./empty');

Object.defineProperty(exports, 'emptyReadStream', {
  enumerable: true,
  get: function get() {
    return _empty.emptyReadStream;
  }
});
Object.defineProperty(exports, 'emptyWriteStream', {
  enumerable: true,
  get: function get() {
    return _empty.emptyWriteStream;
  }
});
Object.defineProperty(exports, 'emptyStreamable', {
  enumerable: true,
  get: function get() {
    return _empty.emptyStreamable;
  }
});

var _pipe = require('./pipe');

Object.defineProperty(exports, 'pipeStream', {
  enumerable: true,
  get: function get() {
    return _pipe.pipeStream;
  }
});

var _quiverStreamChannel = require('quiver-stream-channel');

Object.defineProperty(exports, 'createChannel', {
  enumerable: true,
  get: function get() {
    return _quiverStreamChannel.createChannel;
  }
});
