// quiver-stream-util/out/lib/text.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _quiverPromise = require('quiver-promise');

var _quiverStreamChannel = require('quiver-stream-channel');

var _buffer = require('./buffer');

const streamToText = function streamToText(readStream) {
  return (0, _buffer.streamToBuffer)(readStream).then(function (buffer) {
    return buffer.toString();
  });
};

exports.streamToText = streamToText;
const streamableToText = function streamableToText(streamable) {
  if (streamable.toText) return (0, _quiverPromise.resolve)(streamable.toText());

  return (0, _buffer.streamableToBuffer)(streamable).then(function (buffer) {
    const text = buffer.toString();

    if (streamable.reusable && !streamable.toText) {
      streamable.toText = function () {
        return (0, _quiverPromise.resolve)(text);
      };
    }

    return text;
  });
};

exports.streamableToText = streamableToText;
const textToStream = function textToStream(text) {
  var _createChannel = (0, _quiverStreamChannel.createChannel)();

  const readStream = _createChannel.readStream;
  const writeStream = _createChannel.writeStream;

  writeStream.write(new Buffer(text));
  writeStream.closeWrite();

  return readStream;
};

exports.textToStream = textToStream;
const toTextToStreamable = function toTextToStreamable(toText) {
  let contentType = arguments[1] === undefined ? 'text/plain' : arguments[1];

  let buffer = null;

  const toBuffer = function toBuffer() {
    if (!buffer) buffer = new Buffer(toText());

    return buffer;
  };

  const streamable = (0, _buffer.toBufferToStreamable)(toBuffer);

  streamable.toText = function () {
    return (0, _quiverPromise.resolve)(toText());
  };
  streamable.contentType = contentType;

  return streamable;
};

exports.toTextToStreamable = toTextToStreamable;
const textToStreamable = function textToStreamable(text) {
  let contentType = arguments[1] === undefined ? 'text/plain' : arguments[1];
  return toTextToStreamable(function () {
    return text;
  }, contentType);
};
exports.textToStreamable = textToStreamable;
