// quiver-component/out/lib/router/method-router.js
"use strict";

var _slicedToArray = function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { var _arr = []; for (var _iterator = arr[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) { _arr.push(_step.value); if (i && _arr.length === i) break; } return _arr; } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } };

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var error = require("quiver-error").error;

var _quiverPromise = require("quiver-promise");

var resolve = _quiverPromise.resolve;
var async = _quiverPromise.async;

var emptyStreamable = require("quiver-stream-util").emptyStreamable;

var _quiverHttp = require("quiver-http");

var ResponseHead = _quiverHttp.ResponseHead;
var streamToHttpHandler = _quiverHttp.streamToHttpHandler;

var mapComponent = require("../composite/map").mapComponent;

var HttpHandlerBuilder = require("../http-handler").HttpHandlerBuilder;

var loadHandleable = require("../util/loader").loadHandleable;

const headRequestHandler = function (handler) {
  return function (requestHead, requestStreamable) {
    if (requestHead.method != "HEAD") {
      return handler(requestHead, requestStreamable);
    }

    return handler(requestHead, requestStreamable).then(function (_ref) {
      var _ref2 = _slicedToArray(_ref, 2);

      let responseHead = _ref2[0];
      let responseStreamable = _ref2[1];
      return [responseHead, emptyStreamable()];
    });
  };
};

const optionsHandler = function (methods) {
  const allowedMethods = methods.join(", ") + ", OPTIONS";

  return function (requestHead, requestStreamable) {
    return resolve([new ResponseHead({
      statusCode: 200,
      headers: {
        "content-length": "0",
        allow: allowedMethods
      }
    }), emptyStreamable()]);
  };
};

const methodMapToHttpHandler = function (methodMap) {
  const allowedMethods = Object.keys(methodMap).join(", ");

  const methodNotAllowedResponse = function () {
    return [new ResponseHead({
      statusCode: 405,
      statusMessage: "Method Not Allowed",
      headers: {
        "content-length": "0",
        allow: allowedMethods
      }
    }), emptyStreamable()];
  };

  return function (requestHead, requestStreamable) {
    const method = requestHead.method;
    const handler = methodMap[method];

    if (!handler) return methodNotAllowedResponse();

    return handler(requestHead, requestStreamable);
  };
};

const loadHttpHandler = async(function* (config, component) {
  const handleable = yield loadHandleable(config, component.id, component.builder);

  if (handleable.httpHandler) return handleable.httpHandler;

  if (handleable.streamHandler) return streamToHttpHandler(handleable.streamHandler);

  throw error(500, "handleable is neither stream nor http handler");
});

const loadMethodHandlers = async(function* (config, methodMap) {
  const handlerMap = {};

  for (let key in methodMap) {
    const component = methodMap[key];

    const handler = yield loadHttpHandler(config, component);

    handlerMap[key] = handler;
  }

  if (handlerMap.GET && !handlerMap.HEAD) {
    handlerMap.HEAD = headRequestHandler(handlerMap.GET);
  }

  if (!handlerMap.OPTIONS) {
    handlerMap.OPTIONS = optionsHandler(Object.keys(handlerMap));
  }

  return handlerMap;
});

const methodObjectToMap = function (methodMap) {
  const map = new Map();

  for (let key in methodMap) {
    const handlerComponent = methodMap[key];
    if (!handlerComponent.isHandlerComponent) {
      throw new TypeError("Method map entry value must be handler component");
    }

    map.set(key.toUpperCase(), handlerComponent);
  }

  return map;
};

let MethodRouter = exports.MethodRouter = (function (_HttpHandlerBuilder) {
  function MethodRouter(methodMap) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, MethodRouter);

    options.safeWrapped = true;
    _get(Object.getPrototypeOf(MethodRouter.prototype), "constructor", this).call(this, null, options);

    this.subComponents.methodMap = mapComponent(methodObjectToMap(methodMap));
  }

  _inherits(MethodRouter, _HttpHandlerBuilder);

  _createClass(MethodRouter, {
    toHttpHandlerBuilder: {
      value: function toHttpHandlerBuilder() {
        const snapshot = {};
        const methodMap = this.subComponents.methodMap.map;

        for (let _ref of methodMap.entries()) {
          var _ref2 = _slicedToArray(_ref, 2);

          let key = _ref2[0];
          let component = _ref2[1];

          snapshot[key] = {
            id: component.id,
            builder: component.toHandleableBuilder()
          };
        }

        return function (config) {
          return loadMethodHandlers(config, snapshot).then(methodMapToHttpHandler);
        };
      }
    },
    methodRoute: {
      value: function methodRoute(method, handlerComponent) {
        if (!handlerComponent.isHandlerComponent) {
          throw new TypeError("Method map entry value must be handler component");
        }

        this.subComponents.methodMap.set(method.toUpperCase(), handlerComponent);
      }
    }
  });

  return MethodRouter;
})(HttpHandlerBuilder);

const methodRouter = function (methodMap) {
  return new MethodRouter(methodMap);
};
exports.methodRouter = methodRouter;
