// quiver-component/out/lib/router/route-list-component.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var Component = require("../component").Component;

var listComponent = require("../composite/list").listComponent;

var _routeComponent = require("./route-component");

var Route = _routeComponent.Route;
var StaticRoute = _routeComponent.StaticRoute;
var DynamicRoute = _routeComponent.DynamicRoute;
var RegexRoute = _routeComponent.RegexRoute;
var ParamRoute = _routeComponent.ParamRoute;

var ExtensibleComponent = require("../extensible-component").ExtensibleComponent;

var combineBuilderWithMiddleware = require("../util/middleware").combineBuilderWithMiddleware;

let RouteList = exports.RouteList = (function (_ExtensibleComponent) {
  function RouteList() {
    let options = arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, RouteList);

    _get(Object.getPrototypeOf(RouteList.prototype), "constructor", this).call(this, options);

    this.subComponents.routeList = listComponent();
  }

  _inherits(RouteList, _ExtensibleComponent);

  _createClass(RouteList, {
    routes: {
      get: function () {
        return this.subComponents.routeList.array;
      }
    },
    toRouteBuildSpecs: {
      value: function toRouteBuildSpecs() {
        const middleware = this.toExtendMiddleware();

        return this.routes.map(function (route) {
          const builder = route.toHandleableBuilder();
          const routeSpec = route.routeSpec;

          routeSpec.builder = combineBuilderWithMiddleware(builder, middleware);

          return routeSpec;
        });
      }
    },
    addRoute: {
      value: function addRoute(route) {
        if (!route.isRoute) {
          throw new Error("route must be of type Route");
        }

        this.routes.push(route);

        return this;
      }
    },
    staticRoute: {
      value: function staticRoute(path, handler) {
        return this.addRoute(new StaticRoute(handler, path));
      }
    },
    paramRoute: {
      value: function paramRoute(path, handler) {
        return this.addRoute(new ParamRoute(handler, path));
      }
    },
    regexRoute: {
      value: function regexRoute(regex, fields, handler) {
        return this.addRoute(new RegexRoute(handler, regex, fields));
      }
    },
    dynamicRoute: {
      value: function dynamicRoute(matcher, handler) {
        return this.addRoute(new DynamicRoute(handler, matcher));
      }
    },
    type: {
      get: function () {
        return "route list";
      }
    },
    isRouteList: {
      get: function () {
        return true;
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(RouteList.prototype), "toJson", this).call(this);

        const routes = this.routes;

        if (routes.length > 0) json.routes = this.routes.map(function (route) {
          return route.toJson();
        });

        json.middlewares = this.middlewareJson();

        return json;
      }
    }
  });

  return RouteList;
})(ExtensibleComponent);

const routeList = function (options) {
  return new RouteList(options);
};
exports.routeList = routeList;
