// quiver-component/out/lib/router/route-specs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var copy = require("quiver-object").copy;

var error = require("quiver-error").error;

var _quiverPromise = require("quiver-promise");

var async = _quiverPromise.async;
var reject = _quiverPromise.reject;

var streamToHttpHandler = require("quiver-http").streamToHttpHandler;

const getHandlerFromPath = function (routeIndex, path, args) {
  const staticHandler = routeIndex.staticRoutes[path];
  if (staticHandler) return staticHandler;

  const dynamicRoutes = routeIndex.dynamicRoutes;

  for (let route of dynamicRoutes) {
    const matched = route.matcher(path, args);
    if (matched) return route.handler;
  }

  if (routeIndex.defaultRoute) return routeIndex.defaultRoute;

  return null;
};

const httpRouterHandler = function (routeIndex) {
  return function (requestHead, requestStreamable) {
    const args = requestHead.args;
    const path = requestHead.path;

    const handler = getHandlerFromPath(routeIndex, path, args);
    if (!handler) return reject(error(404, "not found"));

    return handler(requestHead, requestStreamable);
  };
};

const streamRouterHandler = function (routeIndex) {
  return function (args, streamable) {
    var _args$path = args.path;
    const path = _args$path === undefined ? "/" : _args$path;

    const handler = getHandlerFromPath(routeIndex, path, args);
    if (!handler) return reject(error(404, "not found"));

    return handler(args, streamable);
  };
};

const handleableToStreamHandler = function (handleable) {
  return handleable.streamHandler;
};

const handleableToHttpHandler = function (handleable) {
  if (handleable.httpHandler) return handleable.httpHandler;

  if (handleable.streamHandler) return streamToHttpHandler(handleable.streamHandler);

  return null;
};

const routeSpecsToRouteIndex = function (routeSpecs, getHandler) {
  const staticRoutes = {};
  const dynamicRoutes = [];
  let defaultRoute = null;

  routeSpecs.forEach(function (routeSpec) {
    const routeType = routeSpec.routeType;
    const handleable = routeSpec.handleable;

    const handler = getHandler(handleable);
    if (!handler) return;

    if (routeType == "static") {
      const path = routeSpec.path;

      staticRoutes[path] = handler;
    } else if (routeType == "dynamic") {
      const matcher = routeSpec.matcher;

      dynamicRoutes.push({ matcher: matcher, handler: handler });
    } else if (routeType == "default") {
      defaultRoute = handler;
    }
  });

  return {
    staticRoutes: staticRoutes,
    dynamicRoutes: dynamicRoutes,
    defaultRoute: defaultRoute
  };
};

const routeSpecsToRouterHandleable = function (routeSpecs) {
  const streamIndex = routeSpecsToRouteIndex(routeSpecs, handleableToStreamHandler);

  const httpIndex = routeSpecsToRouteIndex(routeSpecs, handleableToHttpHandler);

  const streamHandler = streamRouterHandler(streamIndex);
  const httpHandler = httpRouterHandler(httpIndex);

  return {
    streamHandler: streamHandler,
    httpHandler: httpHandler
  };
};

exports.routeSpecsToRouterHandleable = routeSpecsToRouterHandleable;
const routeBuildSpecsToRouteSpecs = async(function* (config, routeBuildSpecs) {
  const routeSpecs = [];

  for (let i = 0; i < routeBuildSpecs.length; i++) {
    const routeBuildSpec = routeBuildSpecs[i];
    const builder = routeBuildSpec.builder;
    const routeSpec = copy(routeBuildSpec);

    routeSpec.handleable = yield builder(config);
    routeSpecs.push(routeSpec);
  }

  return routeSpecs;
});

const routeBuildSpecsToRouterBuilder = function (routeBuildSpecs) {
  return async(function* (config) {
    const routeSpecs = yield routeBuildSpecsToRouteSpecs(config, routeBuildSpecs);

    return routeSpecsToRouterHandleable(routeSpecs);
  });
};
exports.routeBuildSpecsToRouterBuilder = routeBuildSpecsToRouterBuilder;
