// quiver-component/out/lib/router/route-component.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var async = require("quiver-promise").async;

var loadHandleable = require("../util/loader").loadHandleable;

var combineBuilderWithMiddleware = require("../util/middleware").combineBuilderWithMiddleware;

var _component = require("../component");

var Component = _component.Component;
var HandlerComponent = _component.HandlerComponent;

var MethodRouter = require("./method-router").MethodRouter;

var _dynamicRoute = require("./dynamic-route");

var regexMatcher = _dynamicRoute.regexMatcher;
var paramMatcher = _dynamicRoute.paramMatcher;
var paramUrlBuilder = _dynamicRoute.paramUrlBuilder;

const combineUrlBuilders = function (urlBuilder1, urlBuilder2) {
  if (!urlBuilder1 || !urlBuilder2) return null;

  return function (args) {
    let restPath = arguments[1] === undefined ? "/" : arguments[1];

    const newRestPath = urlBuilder2(args, restPath);

    return urlBuilder1(args, newRestPath);
  };
};

const urlMiddleware = function (urlBuilder) {
  return async(function* (config, builder) {
    const newUrlBuilder = config.urlBuilder = combineUrlBuilders(config.urlBuilder, urlBuilder);

    const handleable = yield builder(config);

    if (!handleable.urlBuilder) {
      handleable.urlBuilder = newUrlBuilder;
    }

    return handleable;
  });
};

const routeBuilder = function (component, urlBuilder, middleware) {
  const mainBuilder = component.toHandleableBuilder();
  const middlewares = [];

  if (urlBuilder) {
    middlewares.push(urlMiddleware(urlBuilder));
  }

  middlewares.push(middleware);

  return combineBuilderWithMiddlewares(mainBuilder, middlewares);
};

let Route = exports.Route = (function (_Component) {
  function Route(handlerComponent) {
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, Route);

    if (!handlerComponent.isHandlerComponent) {
      // argument is method map
      handlerComponent = new MethodRouter(handlerComponent);
    }

    const urlBuilder = options.urlBuilder;

    _get(Object.getPrototypeOf(Route.prototype), "constructor", this).call(this, options);

    this._urlBuilder = urlBuilder;
    this.subComponents.routeHandler = handlerComponent;
  }

  _inherits(Route, _Component);

  _createClass(Route, {
    toHandleableBuilder: {
      value: function toHandleableBuilder() {
        const mainBuilder = this.handlerComponent.toHandleableBuilder();
        const urlBuilder = this.urlBuilder;

        if (urlBuilder) {
          return combineBuilderWithMiddleware(mainBuilder, urlMiddleware(urlBuilder));
        } else {
          return mainBuilder;
        }
      }
    },
    routeSpec: {
      get: function () {
        throw new Error("unimplemented");
      }
    },
    handlerComponent: {
      get: function () {
        return this.subComponents.routeHandler;
      }
    },
    urlBuilder: {
      get: function () {
        return this._urlBuilder;
      }
    },
    type: {
      get: function () {
        return "route";
      }
    },
    isRoute: {
      get: function () {
        return true;
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(Route.prototype), "toJson", this).call(this);
        json.handler = this.handlerComponent.toJson();
        return json;
      }
    }
  });

  return Route;
})(Component);

let StaticRoute = exports.StaticRoute = (function (_Route) {
  function StaticRoute(handlerComponent, staticPath) {
    let options = arguments[2] === undefined ? {} : arguments[2];

    _classCallCheck(this, StaticRoute);

    if (typeof staticPath != "string") throw new TypeError("staticPath must be provided as string");

    const urlBuilder = function () {
      return staticPath;
    };
    options.urlBuilder = options.urlBuilder || urlBuilder;

    _get(Object.getPrototypeOf(StaticRoute.prototype), "constructor", this).call(this, handlerComponent, options);

    this._staticPath = staticPath;
  }

  _inherits(StaticRoute, _Route);

  _createClass(StaticRoute, {
    routeSpec: {
      get: function () {
        return {
          routeType: "static",
          path: this.staticPath
        };
      }
    },
    staticPath: {
      get: function () {
        return this._staticPath;
      }
    },
    type: {
      get: function () {
        return "static route";
      }
    },
    routeType: {
      get: function () {
        return "static";
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(StaticRoute.prototype), "toJson", this).call(this);
        json.staticPath = this.staticPath;
        return json;
      }
    }
  });

  return StaticRoute;
})(Route);

let DynamicRoute = exports.DynamicRoute = (function (_Route2) {
  function DynamicRoute(handlerComponent, matcher) {
    let options = arguments[2] === undefined ? {} : arguments[2];

    _classCallCheck(this, DynamicRoute);

    if (typeof matcher != "function") throw new TypeError("matcher must be of type function");

    _get(Object.getPrototypeOf(DynamicRoute.prototype), "constructor", this).call(this, handlerComponent, options);

    this._matcher = matcher;
  }

  _inherits(DynamicRoute, _Route2);

  _createClass(DynamicRoute, {
    routeSpec: {
      get: function () {
        return {
          routeType: "dynamic",
          matcher: this.matcher
        };
      }
    },
    matcher: {
      get: function () {
        return this._matcher;
      }
    },
    type: {
      get: function () {
        return "dynamic route";
      }
    },
    routeType: {
      get: function () {
        return "dynamic";
      }
    }
  });

  return DynamicRoute;
})(Route);

let RegexRoute = exports.RegexRoute = (function (_DynamicRoute) {
  function RegexRoute(handlerComponent, regex) {
    let matchFields = arguments[2] === undefined ? [] : arguments[2];
    let options = arguments[3] === undefined ? {} : arguments[3];

    _classCallCheck(this, RegexRoute);

    if (!(regex instanceof RegExp)) throw new TypeError("regex must be regular expression");

    const matcher = regexMatcher(regex, matchFields);

    _get(Object.getPrototypeOf(RegexRoute.prototype), "constructor", this).call(this, handlerComponent, matcher, options);

    this._regex = regex;
  }

  _inherits(RegexRoute, _DynamicRoute);

  _createClass(RegexRoute, {
    regex: {
      get: function () {
        return this._regex;
      }
    },
    type: {
      get: function () {
        return "regex route";
      }
    }
  });

  return RegexRoute;
})(DynamicRoute);

let ParamRoute = exports.ParamRoute = (function (_DynamicRoute2) {
  function ParamRoute(handlerComponent, paramPath) {
    let options = arguments[2] === undefined ? {} : arguments[2];

    _classCallCheck(this, ParamRoute);

    if (typeof paramPath != "string") throw new TypeError("param path must be of type string");

    const matcher = paramMatcher(paramPath);

    options.urlBuilder = options.urlBuilder || paramUrlBuilder(paramPath);

    _get(Object.getPrototypeOf(ParamRoute.prototype), "constructor", this).call(this, handlerComponent, matcher, options);

    this._paramPath = paramPath;
  }

  _inherits(ParamRoute, _DynamicRoute2);

  _createClass(ParamRoute, {
    paramPath: {
      get: function () {
        return this._paramPath;
      }
    },
    type: {
      get: function () {
        return "param route";
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(ParamRoute.prototype), "toJson", this).call(this);
        json.paramPath = this.paramPath;
        return json;
      }
    }
  });

  return ParamRoute;
})(DynamicRoute);

const staticRoute = function (handler, path, options) {
  return new StaticRoute(handler, path, options);
};

exports.staticRoute = staticRoute;
const dynamicRoute = function (handler, matcher, options) {
  return new DynamicRoute(handler, matcher, options);
};

exports.dynamicRoute = dynamicRoute;
const regexRoute = function (handler, regex, fields, options) {
  return new RegexRoute(handler, regex, fields, options);
};

exports.regexRoute = regexRoute;
const paramRoute = function (handler, path, options) {
  return new ParamRoute(handler, path, options);
};
exports.paramRoute = paramRoute;
