// quiver-component/out/lib/router/router-component.js
"use strict";

var _toConsumableArray = function (arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } };

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var copy = require("quiver-object").copy;

var async = require("quiver-promise").async;

var normalizeConfig = require("../util/config").normalizeConfig;

var ExtensibleHandler = require("../extensible-component").ExtensibleHandler;

var _component = require("../component");

var Component = _component.Component;
var HandlerComponent = _component.HandlerComponent;

var listComponent = require("../composite/list").listComponent;

var RouteList = require("./route-list-component").RouteList;

var routeBuildSpecsToRouterBuilder = require("./route-specs").routeBuildSpecsToRouterBuilder;

let Router = exports.Router = (function (_ExtensibleHandler) {
  function Router() {
    let options = arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, Router);

    _get(Object.getPrototypeOf(Router.prototype), "constructor", this).call(this, options);

    this.subComponents.defaultRouteList = new RouteList();
    this.subComponents.routeListList = listComponent();
  }

  _inherits(Router, _ExtensibleHandler);

  _createClass(Router, {
    addRoute: {
      value: function addRoute(route) {
        if (!route.isRoute) {
          throw new TypeError("route must be instance of Route");
        }

        this.defaultRouteList.addRoute(route);
        return this;
      }
    },
    staticRoute: {
      value: function staticRoute(path, handler) {
        this.defaultRouteList.staticRoute(path, handler);

        return this;
      }
    },
    paramRoute: {
      value: function paramRoute(path, handler) {
        this.defaultRouteList.paramRoute(path, handler);

        return this;
      }
    },
    regexRoute: {
      value: function regexRoute(regex, fields, handler) {
        this.defaultRouteList.regexRoute(regex, fields, handler);

        return this;
      }
    },
    dynamicRoute: {
      value: function dynamicRoute(matcher, handler) {
        this.defaultRouteList.dynamicRoute(matcher, handler);

        return this;
      }
    },
    routeList: {
      value: (function (_routeList) {
        var _routeListWrapper = function routeList(_x) {
          return _routeList.apply(this, arguments);
        };

        _routeListWrapper.toString = function () {
          return _routeList.toString();
        };

        return _routeListWrapper;
      })(function (routeList) {
        if (!routeList.isRouteList) {
          throw new TypeError("route list must be instance of RouteList");
        }

        this.routeListList.push(routeList);
        return this;
      })
    },
    routeLists: {
      get: function () {
        return [this.defaultRouteList].concat(_toConsumableArray(this.routeListList.array));
      }
    },
    defaultRouteList: {
      get: function () {
        return this.subComponents.defaultRouteList;
      }
    },
    routeListList: {
      get: function () {
        return this.subComponents.routeListList;
      }
    },
    defaultHandler: {
      get: function () {
        return this.subComponents.defaultHandler;
      }
    },
    defaultRoute: {
      value: function defaultRoute(handlerComponent) {
        if (this.defaultHandler) throw new Error("router component already has default route");

        this.subComponents.defaultHandler = handlerComponent;
        return this;
      }
    },
    toMainHandleableBuilder: {
      value: function toMainHandleableBuilder() {
        const routeLists = this.routeLists;

        const routeBuildSpecs = [].concat.apply([], routeLists.map(function (routeList) {
          return routeList.toRouteBuildSpecs();
        }));

        const defaultHandler = this.defaultHandler;

        if (defaultHandler) {
          routeBuildSpecs.push({
            routeType: "default",
            builder: defaultHandler.toHandleableBuilder()
          });
        }

        return routeBuildSpecsToRouterBuilder(routeBuildSpecs);
      }
    },
    type: {
      get: function () {
        return "router";
      }
    },
    toJson: {
      value: function toJson() {
        const json = _get(Object.getPrototypeOf(Router.prototype), "toJson", this).call(this);

        json.routeLists = this.routeLists.map(function (routeList) {
          return routeList.toJson();
        });

        const defaultHandler = this.defaultHandler;
        if (defaultHandler) json.defaultHandler = defaultHandler.toJson();

        json.middlewares = this.middlewareJson();

        return json;
      }
    }
  });

  return Router;
})(ExtensibleHandler);

const router = function (options) {
  return new Router(options);
};

exports.router = router;
const makeRouter = function (options) {
  return new Router(options);
};
exports.makeRouter = makeRouter;
