// quiver-file-stream/out/lib/file-stream.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _slicedToArray(arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i['return']) _i['return'](); } finally { if (_d) throw _e; } } return _arr; } else { throw new TypeError('Invalid attempt to destructure non-iterable instance'); } }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _quiverError = require('quiver-error');

var _quiverPromise = require('quiver-promise');

var _quiverStreamUtil = require('quiver-stream-util');

const nodeFileReadStream = _fs2['default'].createReadStream;
const nodeFileWriteStream = _fs2['default'].createWriteStream;
const unlinkFile = _fs2['default'].unlink;
const existsAsync = _fs2['default'].exists;
const statFile = (0, _quiverPromise.promisify)(_fs2['default'].stat);
exports.statFile = statFile;
const accessFile = (0, _quiverPromise.promisify)(_fs2['default'].access);

exports.accessFile = accessFile;
const isFile = function isFile(fileStats) {
  if (typeof fileStats.isFile == 'function') return fileStats.isFile();

  return fileStats.isFile;
};

const isDirectory = function isDirectory(fileStats) {
  if (typeof fileStats.isDirectory == 'function') return fileStats.isDirectory();

  return fileStats.isDirectory;
};

const getFileStats = function getFileStats(filePath, fileStats) {
  return (fileStats ? (0, _quiverPromise.resolve)(fileStats) : statFile(filePath)).then(function (fileStats) {
    if (!isFile(fileStats)) return reject((0, _quiverError.error)(404, 'file path is not a regular file'));

    return fileStats;
  });
};

const fileExists = function fileExists(filePath) {
  return accessFile(filePath);
};

exports.fileExists = fileExists;
const fileReadStream = function fileReadStream(filePath, fileStats) {
  return getFileStats(filePath, fileStats).then(function () {
    return (0, _quiverStreamUtil.nodeToQuiverReadStream)(nodeFileReadStream(filePath));
  });
};

exports.fileReadStream = fileReadStream;
const fileWriteStream = function fileWriteStream(filePath) {
  return (0, _quiverPromise.resolve)((0, _quiverStreamUtil.nodeToQuiverWriteStream)(nodeFileWriteStream(filePath)));
};

exports.fileWriteStream = fileWriteStream;
/*
 * create a read stream from a temporary file. The temp file
 * is deleted once the read stream piped finish
 */
const tempFileReadStream = function tempFileReadStream(filePath, fileStats) {
  return getFileStats(filePath, fileStats).then(function () {
    const nodeStream = nodeFileReadStream(filePath);

    let deleted = false;
    const deleteFile = function deleteFile() {
      if (deleted) return;

      deleted = true;
      unlinkFile(filePath, function (err) {});
    };

    nodeStream.on('end', deleteFile);
    nodeStream.on('error', deleteFile);

    return (0, _quiverStreamUtil.nodeToQuiverReadStream)(nodeStream);
  });
};

exports.tempFileReadStream = tempFileReadStream;
const streamToFile = function streamToFile(readStream, filePath) {
  return fileWriteStream(filePath).then(function (writeStream) {
    return (0, _quiverStreamUtil.pipeStream)(readStream, writeStream);
  });
};

exports.streamToFile = streamToFile;
/*
 * Create a stream containing ranged content of a file.
 * The start and end values are provided optionally in options.
 * Range convention is the same as Array.slice(), i.e.
 * begin from 0 and not inclusive of end.
 * Example full range is (0, length)
 */
const byteRangeFileStream = function byteRangeFileStream(filePath) {
  let options = arguments[1] === undefined ? {} : arguments[1];
  const fileStats = options.fileStats;
  var _options$start = options.start;
  const start = _options$start === undefined ? 0 : _options$start;
  var _options$end = options.end;
  const end = _options$end === undefined ? -1 : _options$end;

  return getFileStats(filePath, fileStats).then(function (fileStats) {
    const fileSize = fileStats.size;
    if (end == -1) end = fileSize;

    if (fileSize < end) return reject((0, _quiverError.error)(416, 'out of range'));

    return (0, _quiverStreamUtil.nodeToQuiverReadStream)(nodeFileReadStream(filePath, {
      start: start,
      end: end - 1
    }));
  });
};

exports.byteRangeFileStream = byteRangeFileStream;
const fileStreamable = function fileStreamable(filePath, fileStats) {
  return getFileStats(filePath, fileStats).then(function (fileStats) {
    return {
      toStream: function toStream() {
        return (0, _quiverPromise.resolve)(fileReadStream(filePath, fileStats));
      },

      toByteRangeStream: function toByteRangeStream(start, end) {
        return (0, _quiverPromise.resolve)(byteRangeFileStream(filePath, { fileStats: fileStats, start: start, end: end }));
      },

      toFilePath: function toFilePath() {
        return (0, _quiverPromise.resolve)(filePath);
      },

      toNodeStream: function toNodeStream() {
        return (0, _quiverPromise.resolve)(nodeFileReadStream(filePath));
      },

      reusable: true,
      offMemory: true,
      contentLength: fileStats.size
    };
  });
};

exports.fileStreamable = fileStreamable;
/* 
 * Temp file streamable is non-reusable but has file path.
 * Only either toStream() or toFilePath() can be called once.
 * If toStream() is called the temp file is deleted at the end
 * of pipe stream. If toFilePath() is called, it is the caller's
 * responsibility to check for streamable.tempFile flag and delete
 * the file after use.
 */
const tempFileStreamable = function tempFileStreamable(filePath, fileStats) {
  return getFileStats(filePath, fileStats).then(function (fileStats) {
    if (isDirectory(fileStats)) return reject((0, _quiverError.error)(404, 'path is directory'));

    let opened = false;
    const wrap = function wrap(fn) {
      return function () {
        if (opened) return reject((0, _quiverError.error)(500, 'streamable can only be opened once'));

        opened = true;
        return (0, _quiverPromise.resolve)(fn());
      };
    };

    return {
      toStream: wrap(function () {
        return tempFileReadStream(filePath, fileStats);
      }),

      toByteRangeStream: wrap(function () {
        return byteRangeFileStream(filePath, { fileStats: fileStats, start: start, end: end });
      }),

      toFilePath: wrap(function () {
        return filePath;
      }),

      toNodeStream: wrap(function () {
        return nodeFileReadStream(filePath);
      }),

      reusable: false,
      tempFile: true,
      offMemory: true,
      contentLength: fileStats.size
    };
  });
};

exports.tempFileStreamable = tempFileStreamable;
const streamabconstoFile = _asyncToGenerator(function* (streamable, getTempPath) {
  if (streamable.toFilePath) {
    const filePath = yield streamable.toFilePath();
    const isTemp = streamable.tempFile || false;
    return [filePath, isTemp];
  }

  var _ref = yield Promise.all([streamable.toStream(), getTempPath()]);

  var _ref2 = _slicedToArray(_ref, 2);

  const readStream = _ref2[0];
  const tempPath = _ref2[1];

  yield streamToFile(readStream, tempPath);
  return [tempPath, true];
});

exports.streamabconstoFile = streamabconstoFile;
/*
 * Obtain either original or converted streamable that
 * guaranteed to have filepath attribute. A temp file streamable
 * is created if the original streamable has no filePath.
 */
const toFileStreamable = function toFileStreamable(streamable, getTempPath) {
  if (streamable.toFilePath) return (0, _quiverPromise.resolve)(streamable);

  return streamabconstoFile(streamable, getTempPath).then(function (_ref3) {
    var _ref32 = _slicedToArray(_ref3, 1);

    let filePath = _ref32[0];
    return tempFileStreamable(filePath);
  });
};
exports.toFileStreamable = toFileStreamable;
/*ignore*/
