// cloud-firestore/dist/main/firestore/before-opr.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeforeOpr = void 0;
const tharak_query_1 = __importDefault(require("tharak-query"));
class BeforeOpr {
    constructor(providerId) {
        this.providerId = providerId;
    }
    intercept(method, pConfig, input) {
        var oquery = {};
        if (input.filter)
            oquery.$filter = input.filter;
        if (input.select)
            oquery.$select = input.select;
        if (input.update)
            oquery.$update = input.update;
        if (input.insert)
            oquery.$insert = input.insert;
        var options = { dialect: this.providerId };
        if (typeof pConfig.serviceId === 'string')
            options.resource = pConfig.serviceId.trim();
        else {
            options.resource = { name: pConfig.serviceId.body.trim() };
            if (pConfig.serviceId.tag)
                options.resource.alias = pConfig.serviceId.tag;
        }
        let criteria = tharak_query_1.default.build(method, oquery, options);
        if (criteria && criteria.filter && pConfig.input && pConfig.input.params) {
            let idx = 0;
            criteria.filter.forEach((crit) => {
                this.updateRowId(pConfig, crit);
                this.updateObject(pConfig, crit, criteria.filter, idx);
                this.updateArray(pConfig, crit, criteria.filter, idx);
                idx++;
            });
        }
        if (input.select) {
            let selects = input.select.split(',');
            criteria.select = selects;
        }
        return criteria;
    }
    updateRowId(pConfig, crit) {
        let found = pConfig.input.params.find((param) => param.target === crit.field && param.type === "rowId");
        if (found) {
            crit.field = found.type;
        }
    }
    updateObject(pConfig, crit, list, idx) {
        let found = pConfig.input.params.find((param) => param.target === crit.field && param.type === "object");
        if (found) {
            list.splice(idx, 1);
            let obj = JSON.parse(crit.value);
            Object.keys(obj).forEach((key) => {
                list.push({ field: crit.field + '.' + key, operator: '==', value: obj[key] });
            });
        }
    }
    updateArray(pConfig, crit, list, idx) {
        let found = pConfig.input.params.find((param) => param.target === crit.field && param.type === "array");
        if (found) {
            list.splice(idx, 1);
            let array = crit.value;
            if (Array.isArray(array)) {
                array.forEach((key) => {
                    list.push({ field: crit.field, operator: 'array-contains', value: key });
                });
            }
        }
    }
}
exports.BeforeOpr = BeforeOpr;
