// cloud-firestore/dist/main/firestore/store2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Store2 = void 0;
const firebase_admin_1 = __importDefault(require("firebase-admin"));
const rxjs_1 = require("rxjs");
const after_opr_1 = require("./after-opr");
const before_opr_1 = require("./before-opr");
const sql_store_1 = require("./sql-store");
const transaction_1 = require("./transaction");
class Store2 extends sql_store_1.SqlStore {
    constructor() {
        super(...arguments);
        this.observer = new rxjs_1.Observable(observer => {
            observer.next();
            observer.complete();
        });
    }
    onInit(config) {
        if (!config)
            return;
        let adminConfig = {
            credential: firebase_admin_1.default.credential.cert(config.credential),
            databaseURL: config.server
        };
        if (config.appID) {
            firebase_admin_1.default.initializeApp(adminConfig, config.appID);
            let app = firebase_admin_1.default.app(config.appID);
            this.database = firebase_admin_1.default.firestore(app);
            this.beforeOpr = new before_opr_1.BeforeOpr(config.providerId);
            this.providerId = config.providerId;
            this.afterOpr = new after_opr_1.AfterOpr();
            if (config.instanceId)
                this.instanceId = config.instanceId;
        }
        else {
            throw new Error("AppID required.");
        }
    }
    connect(isTransaction, manager) {
        let observable = new rxjs_1.Observable(observer => {
            let currentTrans = manager.findTransaction(this.providerId, this.database);
            if (currentTrans) {
                observer.next(currentTrans);
                observer.complete();
            }
            else {
                if (isTransaction) {
                    let currentTrans = manager.beginTransaction(this.database.batch(), transaction_1.Transaction.REQUIRED_TRANSACTION);
                    observer.next(currentTrans);
                    observer.complete();
                }
                else {
                    observer.next(new transaction_1.Transaction({}, transaction_1.Transaction.NO_TRANSACTION));
                    observer.complete();
                }
            }
        });
        return observable;
    }
    /*connect(isTransaction:boolean, currentTrans): Observable<any> {
        let observable = new Observable<any>(observer => {
            if((!isTransaction && (currentTrans && currentTrans.type === Transaction.NO_TRANSACTION))
            || (currentTrans && currentTrans.type === Transaction.REQUIRED_TRANSACTION)){
                observer.next(currentTrans);
                observer.complete();
            }else{
                if(isTransaction){
                        observer.next(new Transaction(this.database.batch(), Transaction.REQUIRED_TRANSACTION));
                        observer.complete();
                }else{
                    observer.next(new Transaction({}, Transaction.NO_TRANSACTION));
                    observer.complete();
                }
            }
        });
        return observable;
    }*/
    commit(transaction) {
        if (transaction)
            return transaction.commit();
    }
    rollback(transaction) {
        if (transaction)
            return transaction.rollback();
    }
    post(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            let criteria = this.beforeOpr.intercept("add", pConfig, input.body);
            let documentRef = this.database.collection(this.findDBInstance(pConfig.serviceId)).doc();
            let subscriber = () => {
                let result = { opr: 'post' };
                let response = this.afterOpr.intercept(result, pConfig, input.body);
                observer.next([response, transaction]);
                observer.complete();
            };
            if (transaction.type === transaction_1.Transaction.REQUIRED_TRANSACTION) {
                transaction.source.create(documentRef, criteria.value);
                subscriber();
            }
            else {
                this.database.collection(pConfig.serviceId).add(criteria.value).then(subscriber).catch((error) => {
                    observer.error(error);
                });
            }
        });
    }
    put(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            this.inquire(input, pConfig, transaction).subscribe((results) => {
                if (results.length === 3) {
                    let querySnapshot = results[0];
                    let criteria = results[2];
                    querySnapshot.forEach((doc) => {
                        if (transaction.type === transaction_1.Transaction.REQUIRED_TRANSACTION) {
                            transaction.source.update(doc.ref, criteria.value);
                        }
                        else {
                            doc.ref.update(criteria.value);
                        }
                    });
                    let response = this.afterOpr.intercept({ opr: 'put' }, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }
            }, (error) => {
                observer.error(error);
            });
        });
    }
    primaryKey(request, ids, docs) {
        ids.forEach(id => {
            if (typeof id.value === 'string') {
                if (id.value.startsWith('[') && id.value.endsWith(']'))
                    id.value = JSON.parse(id.value);
            }
            if (Array.isArray(id.value)) {
                id.value.forEach(val => {
                    request = request.doc(val);
                    docs.push(request);
                });
            }
            else {
                request = request.doc(id.value);
                docs.push(request);
            }
        });
    }
    evaluate(row, crit) {
        let filter;
        switch (crit.operator) {
            case '==':
                filter = row[crit.field] === crit.value;
                break;
            case '>':
                filter = row[crit.field] > crit.value;
                break;
            case '<':
                filter = row[crit.field] < crit.value;
                break;
            case '<=':
                filter = row[crit.field] <= crit.value;
                break;
            case '>=':
                filter = row[crit.field] >= crit.value;
                break;
            case '<>':
                filter = row[crit.field] !== crit.value;
                break;
        }
        return filter;
    }
    applyIn(filter) {
        let inCrits = filter.filter((crit) => crit.operator === 'in');
        inCrits.forEach();
    }
    addParams(criteria, request) {
        let crits = criteria.filter.filter((crit) => crit.field !== 'rowId' && crit.operator != 'in');
        crits.forEach((crit) => {
            if (crit.field === 'rowId') {
                crit.field = 'id';
                if (crit.operator === 'in') {
                    crit.value.forEach((el) => {
                        request = request.where(crit.field, '==', el);
                    });
                }
            }
            else {
                let fieldPath = crit.field;
                if (crit.field.indexOf('/') > -1) {
                    fieldPath = fieldPath.split('/').join('.');
                }
                if (this.isValidDate(crit.value)) {
                    crit.value = new Date(crit.value);
                }
                request = request.where(`${fieldPath}`, crit.operator, crit.value);
            }
        });
        return request;
    }
    inquire(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            let docs = [];
            let requestBatch = [];
            let proxyRef;
            let criteria = this.beforeOpr.intercept("get", pConfig, input.body);
            let request = this.database.collection(this.findDBInstance(pConfig.serviceId));
            if (criteria && criteria.filter) {
                let ids = criteria.filter.filter((crit) => crit.field === 'rowId');
                if (ids.length > 0)
                    this.primaryKey(request, ids, docs);
                else {
                    let crits = criteria.filter.filter((crit) => crit.operator === 'in');
                    crits.forEach((crit) => {
                        let fieldPath = crit.field;
                        if (crit.field.indexOf('/') > -1) {
                            fieldPath = fieldPath.split('/').join('.');
                        }
                        if (this.isValidDate(crit.value)) {
                            crit.value = new Date(crit.value);
                        }
                        if (Array.isArray(crit.value) && crit.value.length > 9) {
                            let chunks = crit.value.length / 9;
                            if (chunks % 1 > 0) {
                                chunks = Math.floor(chunks);
                                chunks++;
                            }
                            for (let idx = 0; idx < chunks; idx++) {
                                let chunk = crit.value.slice(idx, idx + 9);
                                if (idx > 0)
                                    request = this.database.collection(this.findDBInstance(pConfig.serviceId));
                                request = this.addParams(criteria, request);
                                request = request.where(`${fieldPath}`, crit.operator, chunk);
                                requestBatch.push(request.get());
                            }
                        }
                        else {
                            request = request.where(`${fieldPath}`, crit.operator, crit.value);
                            requestBatch.push(request);
                        }
                    });
                    /*let crits = criteria.filter.filter((crit:any)=>crit.field !== 'rowId');
                    crits.forEach((crit:any)=>{
                        if(crit.field === 'rowId'){
                            crit.field = 'id';
                            if( crit.operator === 'in'){
                                crit.value.forEach((el:any)=>{
                                    request = request.where(crit.field, '==', el);
                                });
                            }
                        }else{
                            let fieldPath = crit.field;
                            if(crit.field.indexOf('/') > -1){
                                fieldPath = fieldPath.split('/').join('.');
                            }
                            if(this.isValidDate(crit.value)){
                                crit.value = new Date(crit.value);
                            }
                            request = request.where(`${fieldPath}`, crit.operator, crit.value);
                        }
                    });*/
                }
            }
            if (docs.length > 0) {
                proxyRef = this.database.getAll.apply(this.database, docs);
            }
            else if (requestBatch.length > 0) {
                proxyRef = Promise.all(requestBatch);
            }
            else {
                if (criteria.select) {
                    proxyRef = request.select.apply(request, criteria.select).get();
                }
                else
                    proxyRef = request.get();
            }
            proxyRef.then((querySnapshot) => {
                if (docs.length > 0) {
                    let crits = criteria.filter.filter((crit) => crit.field !== 'rowId');
                    crits.forEach((crit) => {
                        querySnapshot = querySnapshot.filter((row) => this.evaluate(row.data(), crit));
                    });
                }
                else if (requestBatch.length > 0) {
                    let rstDocs = [];
                    querySnapshot.forEach((querySnapshot) => {
                        querySnapshot.forEach((doc) => {
                            rstDocs.push(doc);
                        });
                    });
                    querySnapshot = rstDocs;
                }
                observer.next([querySnapshot, transaction, criteria]);
                observer.complete();
            }).catch((error) => {
                observer.error(error);
            });
        });
    }
    delete(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            this.inquire(input, pConfig, transaction).subscribe((results) => {
                if (results.length === 3) {
                    let querySnapshot = results[0];
                    querySnapshot.forEach((doc) => {
                        if (transaction.type === transaction_1.Transaction.REQUIRED_TRANSACTION) {
                            transaction.source.delete(doc.ref);
                        }
                        else {
                            doc.ref.delete();
                        }
                    });
                    let response = this.afterOpr.intercept({ opr: 'delete' }, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }
            }, (error) => {
                observer.error(error);
            });
        });
    }
    get(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            if (pConfig.serviceId === 'FIRESTORE.TABLES') {
                let documentRef = this.database.doc(this.findDBInstance());
                documentRef.listCollections().then(collections => {
                    let result = [];
                    for (let collection of collections) {
                        result.push({ TABLE_NAME: collection.id });
                    }
                    let response = this.afterOpr.intercept(result, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }).catch((error) => {
                    observer.error(error);
                });
            }
            else if (pConfig.serviceId === 'FIRESTORE.COLUMNS') {
                let criteria = this.beforeOpr.intercept("get", pConfig, input.body);
                if (criteria && criteria.filter) {
                    let crit = criteria.filter[0];
                    let request = this.database.collection(this.findDBInstance(crit.value));
                    request.get().then((querySnapshot) => {
                        let result = [];
                        let docs = querySnapshot.docs;
                        if (docs.length > 0) {
                            let item = docs[0].data();
                            Object.keys(item).forEach((elem) => {
                                result.push({ "name": elem, "type": "varchar", "instanceId": this.instanceId });
                            });
                        }
                        //let response = this.afterOpr.intercept(result, pConfig, input.body);
                        observer.next([result, transaction]);
                        observer.complete();
                    }).catch((error) => {
                        observer.error(error);
                    });
                }
            }
            else {
                let isSQL = input.headers['sql'];
                let inquiry = this.inquire.bind(this);
                if (isSQL) {
                    inquiry = this.inquire2.bind(this);
                }
                inquiry(input, pConfig, transaction).subscribe((results) => {
                    let result = [];
                    if (results.length > 0) {
                        if (isSQL) {
                            result = results[0];
                        }
                        else {
                            let docRef = results[0];
                            docRef.forEach((doc) => {
                                let item = doc.data();
                                //if(this.isRequireId(input, pConfig))
                                item['rowId'] = doc.id;
                                let foundIdx = result.findIndex((row) => row.rowId === item.rowId);
                                if (foundIdx === -1)
                                    result.push(item);
                            });
                        }
                    }
                    let response = this.afterOpr.intercept(result, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }, (error) => {
                    observer.error(error);
                });
            }
        });
    }
    isRequireId(input, pConfig) {
        if (pConfig.output) {
            if (pConfig.output.isPassthrough) {
                return true;
            }
            else if (pConfig.output.params) {
                let found = pConfig.output.params.find((param) => param.type === 'rowId');
                if (found && input.body.select && input.body.select.indexOf(found.target) > -1) {
                    return true;
                }
            }
        }
        return false;
    }
    isValidDate(dateString) {
        if (typeof dateString === 'string') {
            var regEx = /^\d{4}-\d{2}-\d{2}$/;
            if (!dateString.match(regEx))
                return false; // Invalid format
            var d = new Date(dateString);
            var dNum = d.getTime();
            if (!dNum && dNum !== 0)
                return false; // NaN value, Invalid date
            return d.toISOString().slice(0, 10) === dateString;
        }
        return false;
    }
}
exports.Store2 = Store2;
