// cloud-firestore/dist/main/firestore/store.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Store = void 0;
const firebase_admin_1 = __importDefault(require("firebase-admin"));
const rxjs_1 = require("rxjs");
const after_opr_1 = require("./after-opr");
const before_opr_1 = require("./before-opr");
const sql_store_1 = require("./sql-store");
const transaction_1 = require("./transaction");
class Store extends sql_store_1.SqlStore {
    constructor() {
        super(...arguments);
        this.observer = new rxjs_1.Observable(observer => {
            observer.next();
            observer.complete();
        });
    }
    onInit(config) {
        if (!config)
            return;
        if (config.appID) {
            this.providerId = config.providerId;
            this.appID = config.appID;
            this.adminConfig = {
                credential: firebase_admin_1.default.credential.cert(config.credential),
                databaseURL: config.server
            };
            this.beforeOpr = new before_opr_1.BeforeOpr(config.providerId);
            this.afterOpr = new after_opr_1.AfterOpr();
            if (config.instanceId)
                this.instanceId = config.instanceId;
        }
        else {
            throw new Error("AppID required.");
        }
    }
    /*onInit2(config:any){
        if(!config)
            return;
        let adminConfig = {
            credential: admin.credential.cert(config.credential),
            databaseURL: config.server
            };
        if(config.appID){
            let app = this.findAdminApp(config);
            if(app){
                let initializer = new Observable<any>(observer => {
                    app.delete().then(()=>{
                        console.log("Old App Deleted ");
                        observer.next();
                        observer.complete();
                    }, err => {
                        console.log("error", err);
                    });
                });
                initializer.subscribe(()=>{
                    this.initialize(adminConfig, config);
                });
            }else{
                this.initialize(adminConfig, config);
            }
        }else{
            throw new Error("AppID required.");
        }
    }*/
    findAdminApp() {
        try {
            return firebase_admin_1.default.app(this.appID);
        }
        catch (err) {
            console.log("App not found ");
        }
    }
    initialize() {
        firebase_admin_1.default.initializeApp(this.adminConfig, this.appID);
        let app = firebase_admin_1.default.app(this.appID);
        this.database = firebase_admin_1.default.firestore(app);
    }
    connect(isTransaction, manager) {
        return new rxjs_1.Observable(observer => {
            if (!this.database) {
                let app = this.findAdminApp();
                if (app) {
                    let initializer = new rxjs_1.Observable(iObserver => {
                        app.delete().then(() => {
                            console.log("Old App Deleted ");
                            iObserver.next();
                            iObserver.complete();
                        }, err => {
                            console.log("error", err);
                        });
                    });
                    initializer.subscribe(() => {
                        this.initialize();
                        this.connect2(isTransaction, manager, observer);
                    });
                }
                else {
                    this.initialize();
                    this.connect2(isTransaction, manager, observer);
                }
            }
            else {
                this.connect2(isTransaction, manager, observer);
            }
        });
    }
    connect2(isTransaction, manager, observer) {
        //let observable = new Observable<any>(observer => {
        let currentTrans = manager.findTransaction(this.providerId, this.database);
        if (currentTrans) {
            observer.next(currentTrans);
            observer.complete();
        }
        else {
            if (isTransaction) {
                let currentTrans = manager.beginTransaction(this.database.batch(), transaction_1.Transaction.REQUIRED_TRANSACTION);
                observer.next(currentTrans);
                observer.complete();
            }
            else {
                observer.next(new transaction_1.Transaction({}, transaction_1.Transaction.NO_TRANSACTION));
                observer.complete();
            }
        }
        // });
        //return observable;
    }
    commit(transaction) {
        if (transaction)
            return transaction.commit();
    }
    rollback(transaction) {
        if (transaction)
            return transaction.rollback();
    }
    post(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            let criteria = this.beforeOpr.intercept("add", pConfig, input.body);
            let documentRef = this.database.collection(this.findDBInstance(pConfig.serviceId)).doc();
            let subscriber = (docRef) => {
                let result = { opr: 'post', id: docRef.id };
                let response = this.afterOpr.intercept(result, pConfig, input.body);
                observer.next([response, transaction]);
                observer.complete();
            };
            if (transaction.type === transaction_1.Transaction.REQUIRED_TRANSACTION) {
                transaction.source.create(documentRef, criteria.value);
                subscriber(documentRef);
            }
            else {
                this.database.collection(pConfig.serviceId).add(criteria.value).then(subscriber).catch((error) => {
                    observer.error(error);
                });
            }
        });
    }
    put(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            this.inquire(input, pConfig, transaction).subscribe((results) => {
                if (results.length === 3) {
                    let querySnapshot = results[0];
                    let criteria = results[2];
                    querySnapshot.forEach((doc) => {
                        if (transaction.type === transaction_1.Transaction.REQUIRED_TRANSACTION) {
                            transaction.source.update(doc.ref, criteria.value);
                        }
                        else {
                            doc.ref.update(criteria.value);
                        }
                    });
                    let response = this.afterOpr.intercept({ opr: 'put' }, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }
            }, (error) => {
                observer.error(error);
            });
        });
    }
    primaryKey(request, ids, docs) {
        ids.forEach(id => {
            if (typeof id.value === 'string') {
                if (id.value.startsWith('[') && id.value.endsWith(']'))
                    id.value = JSON.parse(id.value);
            }
            if (Array.isArray(id.value)) {
                id.value.forEach(val => {
                    request = request.doc(val);
                    docs.push(request);
                });
            }
            else {
                request = request.doc(id.value);
                docs.push(request);
            }
        });
    }
    evaluate(row, crit) {
        let filter;
        switch (crit.operator) {
            case '==':
                filter = row[crit.field] === crit.value;
                break;
            case '>':
                filter = row[crit.field] > crit.value;
                break;
            case '<':
                filter = row[crit.field] < crit.value;
                break;
            case '<=':
                filter = row[crit.field] <= crit.value;
                break;
            case '>=':
                filter = row[crit.field] >= crit.value;
                break;
            case '<>':
                filter = row[crit.field] !== crit.value;
                break;
        }
        return filter;
    }
    inquire(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            let docs = [];
            let proxyRef;
            let request = this.database.collection(this.findDBInstance(pConfig.serviceId));
            let criteria = this.beforeOpr.intercept("get", pConfig, input.body);
            if (criteria && criteria.filter) {
                let ids = criteria.filter.filter((crit) => crit.field === 'rowId');
                if (ids.length > 0)
                    this.primaryKey(request, ids, docs);
                else {
                    let crits = criteria.filter.filter((crit) => crit.field !== 'rowId');
                    crits.forEach((crit) => {
                        if (crit.field === 'rowId') {
                            crit.field = 'id';
                            if (crit.operator === 'in') {
                                crit.value.forEach((el) => {
                                    request = request.where(crit.field, '==', el);
                                });
                            }
                        }
                        else {
                            let fieldPath = crit.field;
                            if (crit.field.indexOf('/') > -1) {
                                fieldPath = fieldPath.split('/').join('.');
                            }
                            if (this.isValidDate(crit.value)) {
                                crit.value = new Date(crit.value);
                            }
                            request = request.where(`${fieldPath}`, crit.operator, crit.value);
                        }
                    });
                }
            }
            if (docs.length > 0) {
                proxyRef = this.database.getAll.apply(this.database, docs);
            }
            else {
                if (criteria.select) {
                    proxyRef = request.select.apply(request, criteria.select).get();
                }
                else
                    proxyRef = request.get();
            }
            proxyRef.then((querySnapshot) => {
                if (docs.length > 0) {
                    let crits = criteria.filter.filter((crit) => crit.field !== 'rowId');
                    crits.forEach((crit) => {
                        querySnapshot = querySnapshot.filter((row) => this.evaluate(row.data(), crit));
                    });
                }
                observer.next([querySnapshot, transaction, criteria]);
                observer.complete();
            }).catch((error) => {
                observer.error(error);
            });
        });
    }
    delete(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            this.inquire(input, pConfig, transaction).subscribe((results) => {
                if (results.length === 3) {
                    let querySnapshot = results[0];
                    querySnapshot.forEach((doc) => {
                        if (transaction.type === transaction_1.Transaction.REQUIRED_TRANSACTION) {
                            transaction.source.delete(doc.ref);
                        }
                        else {
                            doc.ref.delete();
                        }
                    });
                    let response = this.afterOpr.intercept({ opr: 'delete' }, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }
            }, (error) => {
                observer.error(error);
            });
        });
    }
    get(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            if (pConfig.serviceId === 'FIRESTORE.TABLES') {
                let documentRef = this.database.doc(this.findDBInstance());
                documentRef.listCollections().then(collections => {
                    let result = [];
                    for (let collection of collections) {
                        result.push({ TABLE_NAME: collection.id });
                    }
                    let response = this.afterOpr.intercept(result, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }).catch((error) => {
                    observer.error(error);
                });
            }
            else if (pConfig.serviceId === 'FIRESTORE.COLUMNS') {
                let criteria = this.beforeOpr.intercept("get", pConfig, input.body);
                if (criteria && criteria.filter) {
                    let crit = criteria.filter[0];
                    let request = this.database.collection(this.findDBInstance(crit.value));
                    request.get().then((querySnapshot) => {
                        let result = [];
                        let docs = querySnapshot.docs;
                        if (docs.length > 0) {
                            let item = docs[0].data();
                            Object.keys(item).forEach((elem) => {
                                result.push({ "name": elem, "type": "varchar", "instanceId": this.instanceId });
                            });
                        }
                        //let response = this.afterOpr.intercept(result, pConfig, input.body);
                        observer.next([result, transaction]);
                        observer.complete();
                    }).catch((error) => {
                        observer.error(error);
                    });
                }
            }
            else {
                let isSQL = input.headers['sql'];
                let inquiry = this.inquire.bind(this);
                if (isSQL) {
                    inquiry = this.inquire2.bind(this);
                }
                inquiry(input, pConfig, transaction).subscribe((results) => {
                    let result = [];
                    if (results.length > 0) {
                        if (isSQL) {
                            result = results[0];
                        }
                        else {
                            let docRef = results[0];
                            docRef.forEach((doc) => {
                                let item = doc.data();
                                if (this.isRequireId(input, pConfig))
                                    item['rowId'] = doc.id;
                                result.push(item);
                            });
                        }
                    }
                    let response = this.afterOpr.intercept(result, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }, (error) => {
                    observer.error(error);
                });
            }
        });
    }
    isRequireId(input, pConfig) {
        if (pConfig.output) {
            if (pConfig.output.isPassthrough) {
                return true;
            }
            else if (pConfig.output.params) {
                let found = pConfig.output.params.find((param) => param.type === 'rowId');
                if (found && input.body.select && input.body.select.indexOf(found.target) > -1) {
                    return true;
                }
            }
        }
        return false;
    }
    isValidDate(dateString) {
        if (typeof dateString === 'string') {
            var regEx = /^\d{4}-\d{2}-\d{2}$/;
            if (!dateString.match(regEx))
                return false; // Invalid format
            var d = new Date(dateString);
            var dNum = d.getTime();
            if (!dNum && dNum !== 0)
                return false; // NaN value, Invalid date
            return d.toISOString().slice(0, 10) === dateString;
        }
        return false;
    }
}
exports.Store = Store;
