// cloud-firestore/dist/main/firestore/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
class Transaction {
    constructor(source, type) {
        this.source = source;
        this.type = type;
        if (this.type === Transaction.REQUIRED_TRANSACTION)
            console.log('Transaction Started');
    }
    getType() {
        return this.type;
    }
    commit() {
        if (this.type === Transaction.REQUIRED_TRANSACTION) {
            this.source.commit().then(() => {
                console.log('Transaction Committed');
            }).catch(err => {
                console.log('Transaction Error ' + err.toString());
            });
        }
    }
    rollback() {
        if (this.type === Transaction.REQUIRED_TRANSACTION) {
            this.source.rollback().then(() => {
                console.log('Transaction Rolledback');
            });
        }
    }
    static commit(transaction) {
        if (transaction)
            return transaction.commit();
    }
    static rollback(transaction) {
        if (transaction)
            return transaction.rollback();
    }
}
exports.Transaction = Transaction;
Transaction.REQUIRED_TRANSACTION = 1;
Transaction.NO_TRANSACTION = 0;
