// cloud-firestore/src/main/firestore/store.ts
import admin from 'firebase-admin';
import { Observable } from 'rxjs';
import { AfterOpr } from './after-opr';
import { BeforeOpr } from './before-opr';
import { SqlStore } from './sql-store';
import { Transaction } from './transaction';
export class Store extends SqlStore{
    observer:Observable<any> = new Observable<any>(observer => {            
        observer.next();
        observer.complete();
    });
    providerId:string;
    appID:string;
    adminConfig:any;
    onInit(config:any){
        if(!config)
            return;
        if(config.appID){
            this.providerId = config.providerId;
            this.appID = config.appID;            
            this.adminConfig = {
                credential: admin.credential.cert(config.credential),
                databaseURL: config.server
                };
            this.beforeOpr = new BeforeOpr(config.providerId);
            this.afterOpr = new AfterOpr();
            if(config.instanceId)
                this.instanceId = config.instanceId; 
        }else{
            throw new Error("AppID required.");       
        }        
    }
    /*onInit2(config:any){
        if(!config)
            return;
        let adminConfig = {
            credential: admin.credential.cert(config.credential),
            databaseURL: config.server
            };
        if(config.appID){
            let app = this.findAdminApp(config);
            if(app){
                let initializer = new Observable<any>(observer => {
                    app.delete().then(()=>{
                        console.log("Old App Deleted ");
                        observer.next();
                        observer.complete();                        
                    }, err => {
                        console.log("error", err);
                    });
                });
                initializer.subscribe(()=>{
                    this.initialize(adminConfig, config);
                });
            }else{
                this.initialize(adminConfig, config);
            }            
        }else{
            throw new Error("AppID required.");       
        }
    }*/
    private findAdminApp(){
        try{
            return admin.app(this.appID);
        }catch(err){
            console.log("App not found ");
        }
    }
    private initialize(){
        admin.initializeApp(this.adminConfig, this.appID);
        let app = admin.app(this.appID);
        this.database = admin.firestore(app);         
    }
    connect(isTransaction:boolean, manager:any): Observable<any> {
        return new Observable<any>(observer => {
            if(!this.database){
                let app = this.findAdminApp();
                if(app){
                    let initializer = new Observable<any>(iObserver => {
                        app.delete().then(()=>{
                            console.log("Old App Deleted ");
                            iObserver.next();
                            iObserver.complete();                        
                        }, err => {
                            console.log("error", err);
                        });
                    });
                    initializer.subscribe(()=>{
                        this.initialize();
                        this.connect2(isTransaction, manager, observer);
                    });
                }else{
                    this.initialize();
                    this.connect2(isTransaction, manager, observer);
                }
            }else{
                this.connect2(isTransaction, manager, observer);
            }
        });
    }
    private connect2(isTransaction:boolean, manager:any, observer:any){
        //let observable = new Observable<any>(observer => {
            let currentTrans = manager.findTransaction(this.providerId, this.database);
            if(currentTrans){
                observer.next(currentTrans);
                observer.complete();
            }else{
                if(isTransaction){
                    let currentTrans = manager.beginTransaction(this.database.batch(), Transaction.REQUIRED_TRANSACTION);
                    observer.next(currentTrans);
                    observer.complete();                    
                }else{
                    observer.next(new Transaction({}, Transaction.NO_TRANSACTION));
                    observer.complete();
                }
            }
       // });
        //return observable;
    }    

    commit(transaction:any){
        if(transaction)
            return transaction.commit();
    }
    rollback(transaction:any){
        if(transaction)
            return transaction.rollback();
    }
    post(input:any, pConfig:any, transaction:any) {
        return new Observable<any>(observer => { 
            let criteria = this.beforeOpr.intercept("add", pConfig, input.body);
            let documentRef = this.database.collection(this.findDBInstance(pConfig.serviceId)).doc();
            let subscriber = (docRef:any)=>{
                let result = {opr : 'post', id: docRef.id};
                let response = this.afterOpr.intercept(result, pConfig, input.body);
                observer.next([response, transaction]);
                observer.complete();
            };
            if(transaction.type === Transaction.REQUIRED_TRANSACTION){
                transaction.source.create(documentRef, criteria.value);
                subscriber(documentRef);
            }else{
                this.database.collection(pConfig.serviceId).add(criteria.value).then(subscriber).catch((error) => {
                    observer.error(error);
                  });
            }
        });
    }
    put(input:any, pConfig:any, transaction:any) {
        return new Observable<any>(observer => {
            this.inquire(input, pConfig, transaction).subscribe((results:any)=>{
                if(results.length === 3){
                    let querySnapshot = results[0];
                    let criteria = results[2];
                    querySnapshot.forEach((doc)=>{
                        if(transaction.type === Transaction.REQUIRED_TRANSACTION){
                            transaction.source.update(doc.ref, criteria.value);
                        }else{
                            doc.ref.update(criteria.value);
                        }            
                    });
                        let response = this.afterOpr.intercept({opr: 'put'}, pConfig, input.body);
                        observer.next([response, transaction]);
                        observer.complete();
                }
            }, (error)=>{
                observer.error(error);
            });            
        });        
    }
    
    private primaryKey(request:any, ids:any, docs:any []){
        ids.forEach(id=>{
            if(typeof id.value === 'string'){
                if(id.value.startsWith('[') && id.value.endsWith(']'))
                    id.value = JSON.parse(id.value);
            }
            if(Array.isArray(id.value)){
                id.value.forEach(val=>{
                    request = request.doc(val);
                    docs.push(request);
                });
            }else{
                request = request.doc(id.value);
                docs.push(request);
            }
        });
    }
    private evaluate(row:any, crit:any){
        let filter;
        switch(crit.operator){
            case '==':
                filter = row[crit.field] === crit.value;
                break;
            case '>':
                filter = row[crit.field] > crit.value;
                break;
            case '<':
                filter = row[crit.field] < crit.value;
                break;
            case '<=':
                filter = row[crit.field] <= crit.value;
                break;
            case '>=':
                filter = row[crit.field] >= crit.value;
                break;
            case '<>':
                filter = row[crit.field] !== crit.value;
                break;                                                                    
        }
        return filter;
    }

    inquire(input:any, pConfig:any, transaction:any) {
        return new Observable<any>(observer => { 
            let docs = [];
            let proxyRef;
            let request = this.database.collection(this.findDBInstance(pConfig.serviceId));
            let criteria = this.beforeOpr.intercept("get", pConfig, input.body);
            if(criteria && criteria.filter){
                let ids = criteria.filter.filter((crit:any)=>crit.field === 'rowId');
                if(ids.length > 0)
                    this.primaryKey(request, ids, docs);
                else{
                    let crits = criteria.filter.filter((crit:any)=>crit.field !== 'rowId');
                    crits.forEach((crit:any)=>{
                        if(crit.field === 'rowId'){
                            crit.field = 'id';
                            if( crit.operator === 'in'){
                                crit.value.forEach((el:any)=>{                                
                                    request = request.where(crit.field, '==', el);
                                });
                            }
                        }else{
                            let fieldPath = crit.field;
                            if(crit.field.indexOf('/') > -1){
                                fieldPath = fieldPath.split('/').join('.');
                            }
                            if(this.isValidDate(crit.value)){
                                crit.value = new Date(crit.value);
                            }
                            request = request.where(`${fieldPath}`, crit.operator, crit.value);
                        }
                    });
                }
            }
            if(docs.length > 0){
                proxyRef = this.database.getAll.apply(this.database, docs);
            }else{
                if(criteria.select){                    
                    proxyRef = request.select.apply(request, criteria.select).get();
                }else
                    proxyRef = request.get();
            }
            proxyRef.then((querySnapshot)=>{                
                if(docs.length > 0){
                    let crits = criteria.filter.filter((crit:any)=>crit.field !== 'rowId');
                    crits.forEach((crit:any)=>{
                        querySnapshot = querySnapshot.filter((row:any)=>this.evaluate(row.data(), crit));
                    });
                }
                observer.next([querySnapshot, transaction, criteria]);
                observer.complete();
            }).catch((error) => {
                observer.error(error);
              });            
        });
    }  
    delete(input:any, pConfig:any, transaction:any) {
        return new Observable<any>(observer => {           
            this.inquire(input, pConfig, transaction).subscribe((results:any)=>{                
                if(results.length === 3){
                    let querySnapshot = results[0];
                    querySnapshot.forEach((doc)=>{
                        if(transaction.type === Transaction.REQUIRED_TRANSACTION){
                            transaction.source.delete(doc.ref);
                        }else{
                            doc.ref.delete();  
                        }         
                    });
                        let response = this.afterOpr.intercept({opr: 'delete'}, pConfig, input.body);
                        observer.next([response, transaction]);
                        observer.complete();
                }
            }, (error)=>{
                observer.error(error);
            });            
        });        
    }
    get(input:any, pConfig:any, transaction:any) {
        return new Observable<any>(observer => {
            if(pConfig.serviceId === 'FIRESTORE.TABLES'){
                let documentRef = this.database.doc(this.findDBInstance());
                documentRef.listCollections().then(collections => {
                    let result = [];
                    for (let collection of collections) {
                        result.push({TABLE_NAME: collection.id});
                    }
                    let response = this.afterOpr.intercept(result, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }).catch((error) => {
                    observer.error(error);
                  });
            }else if(pConfig.serviceId === 'FIRESTORE.COLUMNS'){
                let criteria = this.beforeOpr.intercept("get", pConfig, input.body);
                if(criteria && criteria.filter){
                    let crit = criteria.filter[0];
                    let request = this.database.collection(this.findDBInstance(crit.value));
                    request.get().then((querySnapshot)=>{
                        let result = [];
                        let docs = querySnapshot.docs;
                        if(docs.length > 0){                            
                            let item = docs[0].data();
                            Object.keys(item).forEach((elem:any)=>{
                                result.push({"name":elem,"type":"varchar","instanceId":this.instanceId});
                            });
                        }
                        //let response = this.afterOpr.intercept(result, pConfig, input.body);
                        observer.next([result, transaction]);
                        observer.complete();
                    }).catch((error) => {
                        observer.error(error);
                      });
                }
            }else{                            
                let isSQL = input.headers['sql'];
                let inquiry = this.inquire.bind(this);
                if(isSQL){
                    inquiry = this.inquire2.bind(this);
                }
                inquiry(input, pConfig, transaction).subscribe((results:any)=>{
                    let result = [];
                    if(results.length > 0){
                        if(isSQL){
                            result = results[0];
                        }else{
                            let docRef = results[0];
                            docRef.forEach((doc)=>{
                                let item = doc.data();
                                if(this.isRequireId(input, pConfig))
                                item['rowId'] = doc.id;
                                result.push(item);
                            });
                        }                    
                    }
                    let response = this.afterOpr.intercept(result, pConfig, input.body);
                    observer.next([response, transaction]);
                    observer.complete();
                }, (error)=>{
                    observer.error(error);
                });
            }
        });
    }
    isRequireId(input:any, pConfig:any):boolean{
        if(pConfig.output){
            if(pConfig.output.isPassthrough){
                return true;
            }else if(pConfig.output.params){
                let found = pConfig.output.params.find((param:any)=>param.type === 'rowId');
                if(found && input.body.select && input.body.select.indexOf(found.target)> -1){
                    return true;
                }
            }
        }
        return false;
    }
    isValidDate(dateString) {
        if(typeof dateString === 'string'){
            var regEx = /^\d{4}-\d{2}-\d{2}$/;
            if(!dateString.match(regEx)) return false;  // Invalid format
            var d = new Date(dateString);
            var dNum = d.getTime();
            if(!dNum && dNum !== 0) return false; // NaN value, Invalid date
            return d.toISOString().slice(0,10) === dateString;
        }
        return false;
      }
}
