// cloud-firestore/src/main/firestore/transaction.ts
export class Transaction {
    static REQUIRED_TRANSACTION:Number = 1;
    static NO_TRANSACTION:Number = 0;
    source:any;
    type:Number
    constructor(source:any, type:Number){
        this.source = source;
        this.type = type;
        if(this.type === Transaction.REQUIRED_TRANSACTION)
            console.log('Transaction Started');
    }
    getType(){
        return this.type;
    }
    commit(){
        if(this.type === Transaction.REQUIRED_TRANSACTION){
            this.source.commit().then(()=>{
                console.log('Transaction Committed');
            }).catch(err=>{
                console.log('Transaction Error '+err.toString());
            });
        }
    }
    rollback(){
        if(this.type === Transaction.REQUIRED_TRANSACTION){
            this.source.rollback().then(()=>{
                console.log('Transaction Rolledback');
            });
        }
    }
    static commit(transaction:any){
        if(transaction)
        return transaction.commit();
    }
    static rollback(transaction:any){
        if(transaction)
        return transaction.rollback();
    }
}
