// quiver-object/out/lib/copy.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _ownKeys = require('./keys');

const defaultProto = Object.getPrototypeOf({});

const noCopyKey = Symbol('noCopy');

const shouldNotCopy = function shouldNotCopy(object) {
  return object[noCopyKey] == true;
};

const copy = function copy(object) {
  let options = arguments[1] === undefined ? {} : arguments[1];
  var _options$forceCopy = options.forceCopy;
  const forceCopy = _options$forceCopy === undefined ? false : _options$forceCopy;
  const excludeFields = options.excludeFields;

  if (object == null || typeof object != 'object') {
    return object;
  }

  if (shouldNotCopy(object) && !forceCopy) {
    return object;
  }

  if (Array.isArray(object)) {
    return copyArray(object);
  }

  /*
   * We don't perform copy on the object if it is not a plain object,
   * i.e. has a different prototype than the default one.
   *
   * The evaluation is the same as
   *  Object.getPrototypeOf(object) === Object.getPrototypeOf({})
   * Except that we cache the default prototype in the constiable defaultProto.
   *
   * This works because all objects created through the hash literal
   * or new Object() all shares the same prototype.
   */
  const proto = Object.getPrototypeOf(object);
  if (proto !== defaultProto && proto != null) {
    return object;
  }

  if (excludeFields) {
    return copyPlainObjectWithExcludes(object, excludeFields);
  } else {
    return copyPlainObject(object);
  }
};

exports.copy = copy;
const copyPlainObject = function copyPlainObject(object) {
  const newObject = Object.create(null);

  for (let key of _ownKeys.ownKeys(object)) {
    newObject[key] = copy(object[key]);
  }

  return newObject;
};

const copyPlainObjectWithExcludes = function copyPlainObjectWithExcludes(object, excludeFields) {
  const newObject = Object.create(null);

  for (let key in object) {
    if (excludeFields.indexOf(key) > -1) continue;

    const value = object[key];
    newObject[key] = copy(value);
  }

  return newObject;
};

const copyArray = function copyArray(array) {
  const newArray = [];
  array.forEach(function (value) {
    newArray.push(copy(value));
  });
  return newArray;
};

const noCopy = function noCopy(obj) {
  if (!obj) {
    return obj;
  }Object.defineProperty(obj, noCopyKey, {
    enumerable: false,
    writable: true,
    value: true
  });
  return obj;
};
exports.noCopy = noCopy;
