// quiver-object/out/lib/object.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _merge = require('./merge');

Object.defineProperty(exports, 'merge', {
  enumerable: true,
  get: function get() {
    return _merge.merge;
  }
});

var _mixin = require('./mixin');

Object.defineProperty(exports, 'mixin', {
  enumerable: true,
  get: function get() {
    return _mixin.mixin;
  }
});

var _ownKeys = require('./keys');

Object.defineProperty(exports, 'ownKeys', {
  enumerable: true,
  get: function get() {
    return _ownKeys.ownKeys;
  }
});

var _copy$noCopy = require('./copy');

Object.defineProperty(exports, 'copy', {
  enumerable: true,
  get: function get() {
    return _copy$noCopy.copy;
  }
});
Object.defineProperty(exports, 'noCopy', {
  enumerable: true,
  get: function get() {
    return _copy$noCopy.noCopy;
  }
});

var _defineGetter = require('./getter');

Object.defineProperty(exports, 'defineGetter', {
  enumerable: true,
  get: function get() {
    return _defineGetter.defineGetter;
  }
});

var _assertUndefined$assertInstanceOf$assertArrayInstanceOf$assertFunction$assertString = require('./assert');

Object.defineProperty(exports, 'assertUndefined', {
  enumerable: true,
  get: function get() {
    return _assertUndefined$assertInstanceOf$assertArrayInstanceOf$assertFunction$assertString.assertUndefined;
  }
});
Object.defineProperty(exports, 'assertInstanceOf', {
  enumerable: true,
  get: function get() {
    return _assertUndefined$assertInstanceOf$assertArrayInstanceOf$assertFunction$assertString.assertInstanceOf;
  }
});
Object.defineProperty(exports, 'assertArrayInstanceOf', {
  enumerable: true,
  get: function get() {
    return _assertUndefined$assertInstanceOf$assertArrayInstanceOf$assertFunction$assertString.assertArrayInstanceOf;
  }
});
Object.defineProperty(exports, 'assertFunction', {
  enumerable: true,
  get: function get() {
    return _assertUndefined$assertInstanceOf$assertArrayInstanceOf$assertFunction$assertString.assertFunction;
  }
});
Object.defineProperty(exports, 'assertString', {
  enumerable: true,
  get: function get() {
    return _assertUndefined$assertInstanceOf$assertArrayInstanceOf$assertFunction$assertString.assertString;
  }
});
