// cloud-firestore/src/test/test.ts
import { Store } from "../main/firestore/store";
import * as clientConfig from './resources/test-config.json';
//import express from 'express';
//import bodyParser from 'body-parser';
import { Util } from "./util";
import { Uploader } from "./uploader";
import { SqlStore } from "../main/firestore/sql-store";
import { TestSql } from "./test-sql";
import { Store2 } from "../main/firestore/store2";
export class Test {
    static start(){
        /*const app = express();  
        app.use(bodyParser.json({
            limit: '50mb'
          }));
      
          app.use(bodyParser.urlencoded({
            limit: '50mb',
            parameterLimit: 100000,
            extended: true 
          }));
          let args = this.getArgs();         
          var serverPort = process.env.PORT || args['port'];  
          app.listen(serverPort, ()=>{  return console.log(`server is listening on ${serverPort}`);
          });*/
          let provider = new Store2();         
              provider.onInit({
                "appID" : "adminStore",
                "credential" : {
                    "type": "service_account",
                    "project_id": "mydir-fd30b",
                    "private_key_id": "10687391cd57058d4f6bc31e5e08e06a1b277677",
                    "private_key": "-----BEGIN PRIVATE KEY-----\nMIIEvAIBADANBgkqhkiG9w0BAQEFAASCBKYwggSiAgEAAoIBAQDKKlYWSooBsuky\nleGF3x8P7+igdUpwJxRy1urugirlnTPJ1NHUkpOd+2SFyO6MrzoH5p06adt3mCgR\n6eH/C3Ll8PMKq4gxJkVXDT1LwBwZlVmPwmTamSBq7qbiZ3ERilUBzVPCD6u3PhDN\nVjEXUxW1un7FjKr2apY/sPXigXrJvd+b9O4dLKXZiRXcFvbYGqt2A5ojOAvRRa0N\nUJytXtCnDxgSVRsvvuIEjhwcaSTIzX8IsCoDYL5umq0b3lC2ElL5HufNxapaGT7I\nfgA4vJDHyxrbjVjzNkmr3FfWE8GpzS5NR8oDsm2zOQZvZ8QjvmXhvYXDAZDkzyhl\nJPkDa0djAgMBAAECggEAGcUfyCOkR+codrxITSpN6VgpjRqmbHluz9y5gcGx+ib5\nzQaF4NT2YDMFc1BTqDdF7G/nJuULhsksgmwP1NT+xnTaZkfLSoiYWipj394TNuek\nzQB4nqzaRwc42q9Y6GV7xIkJ3jjlKKyOs74C7/g1Uj79nez6dgLMOD40ebULk9+Y\nNzSYvG4Qrerm4j7ajy/fV8/4ORS63V1AVsM7cFQREqUtNoyC1mydXhFTiSLrNm4q\nqLC+9q9pSMgMXi8sOsSZNMezxPt998f+yjxOIx2f259abVsvFbvaHjWxGkz3G7Yk\nk+nSxK8DngOCa/ZyBJA6MyWfGxapnDlXxEq6WNbN0QKBgQDpMWR4LfcGR6kXDXkm\nCYLU2a4EsLYqkoqYSWg0zo38BFN9pQX2vfjoK/mwYxWmcUXn9k5XF62zb+8hl7x4\n33J1DeCKeqURKnObxcsvteCqXTgKVY3G9cXOTvZkXyG2TeuRtN+kWxK6n/3k1BGI\nqW6V1XnAyydeigfXiC/GaheJrQKBgQDd8BYCUFOPJSmzQ95TxzmvBz3dPqDkrJ3b\nF8NSAQeUQu6ZZLy+fmpbVZYGxLY3whE1qK21hRd+s14aXxEo/EqOBLYqb0f6OjCP\nJWpMoxqRXyNGEUGvv3LQvDT462P/NAF4/DuR4LWCQiSgD2PHD5oSA1J4hTCHs53y\nXkKlZ5ZXTwKBgCTLk2PXNOQ7yHWHvbb/xzIYyvmYfJYJG29eHeA/O24prJblxCQj\n47OWxMZNTCxAMVyho8jgRptKrTvqumWENIY901FmesdToHxS4HW9yB55qQ8QYtej\nDjw4f25lCGfAscwC80r2x+TEvAYwZDfhLD7eOWDvbV3gf85FgexX7zDdAoGAbFDN\nXQm5Dkaz6GK6MicsVujw33a+bXx1ekVJtU79VPWpn0GUAGCQJc1KcDk3OVVEsuyF\nAu9WvSZrSfGV3WvsMw3HnkQ7W7tkkahQDeEYiVRmUb8AUPbd4VPJgqbXYekOwu5P\nYG3MCCuh1mBMcQ2DQ0FN0bYnLkOSVYR9gF5i+G8CgYBSVpXEFY50SiQbOrAg7DTu\neaS6S8i/8cnHQtOx/n66oibRCSTnztWEXsRdAUZ2SdA+cdvF3ecWJ6GwkPVTIr0i\nXzNVKb0lujuJmXdUc6iSKevFDoWtlBCiY9Q6/DZkzbTh0BYH795rbFSbpMkjxA03\n+b7SsXAzL5dq3PI6FfUsDw==\n-----END PRIVATE KEY-----\n",
                    "client_email": "firebase-adminsdk-rfxpy@mydir-fd30b.iam.gserviceaccount.com",
                    "client_id": "108683160193367885510",
                    "auth_uri": "https://accounts.google.com/o/oauth2/auth",
                    "token_uri": "https://oauth2.googleapis.com/token",
                    "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
                    "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/firebase-adminsdk-rfxpy%40mydir-fd30b.iam.gserviceaccount.com"
                  },
                "instanceId" : "EMR_STORE/0YcKrOiXReRPhISfQXjh",
                "providerId" : "firestore",
                "server": "https://mydir-fd30b.firebaseio.com",
                "storageBucket": "mydir-fd30b.appspot.com"
              });        
          /*app.get('/store', (req, res)=>{
            let sConfig:any = {'serviceId': "TEST_POST", 'input': {}, 'output': {'isPassthrough': true}, 'oprType': {}, 'target': 'adminStore', Util: Util};
              provider.connect(false, {
                findTransaction: function(){
                  return;
                }
              }).subscribe((cmd)=>{
                provider.get({body:{"select":"NAME","filter":"NAME eq '[\"Fever\",\"Fever\",\"Fever\",\"Fever\",\"Fever\",\"Fever\",\"Fever\",\"Fever\",\"Fever\",\"Fever\",\"Fever\"]'"}, headers:{}}, sConfig, cmd).subscribe(result=>{
                    res.send(JSON.stringify(result[0]));
                    res.status(200);
                    res.end();
                  });
              }, err=>{
                res.send(err.toString());
                    res.status(200);
                    res.end();
              });
        });*/
    }
    static getArgs () {
        const args = {};
        process.argv
            .slice(2, process.argv.length)
            .forEach( arg => {
            // long arg
            if (arg.slice(0,2) === '--') {
                const longArg = arg.split('=');
                const longArgFlag = longArg[0].slice(2,longArg[0].length);
                const longArgValue = longArg.length > 1 ? longArg[1] : true;
                args[longArgFlag] = longArgValue;
            }
            // flags
            else if (arg[0] === '-') {
                const flags = arg.slice(1,arg.length).split('');
                flags.forEach(flag => {
                args[flag] = true;
                });
            }
        });
        return args;
      }
}
Test.start();
/*let uploader = new Uploader();
uploader.load();
uploader.save();*/
/*let testSql = new TestSql();
testSql.load();
testSql.createServer();*/
