// cloud-firestore/src/test/test-sql.ts
import { Store } from "../main/firestore/store";
import { Util } from "./util";
//import express from 'express';
//import bodyParser from 'body-parser';
import { PatientLoader } from "./patient-loader";

export class TestSql {
    provider:any = new Store();
    emrLoader:PatientLoader = new PatientLoader();
    load(){
      this.emrLoader.load(this.provider);             
    }
    createServer(){
        /*const app = express();  
        app.use(bodyParser.json({
            limit: '50mb'
          }));
      
          app.use(bodyParser.urlencoded({
            limit: '50mb',
            parameterLimit: 100000,
            extended: true 
          }));
          let args = this.getArgs();         
          var serverPort = process.env.PORT || args['port'];  
          app.listen(serverPort, ()=>{  return console.log(`server is listening on ${serverPort}`);
          });
          this.query(app);*/
    }
    private query(app:any){
        let sConfig:any = {'serviceId': "TM_APPOINTMENTS", 'input': {}, 'output': {'isPassthrough': true}, 'oprType': {}, 'target': 'adminStore', Util: Util};
        app.get('/load', (req, res)=>{
          this.provider.connect(true, null).subscribe((cmd)=>{            
            //let reqObj:any = {filter: "DOCTCODE eq '000001' and ((APPOINTMENTDATE ge 1598832000000) and APPOINTMENTDATE le 1600905600000)"};//{"filter":"mobile_code ne null"};            
            let reqObj:any = {select: "VISITSTATUS,DOCTCODE,APPOINTMENTDATE,APPOINTMENTNO,MOBILE", filter: "DOCTCODE eq '000001' and ((APPOINTMENTDATE ge '2020-08-01') and APPOINTMENTDATE le '2020-11-01')"};          
                this.provider.get({body:reqObj, headers:{}}, sConfig, cmd).subscribe(result=>{                    
                    cmd.commit();
                    res.send(JSON.stringify(result[0]));
                    res.status(200);
                    res.end();
                    
                });
            });
        });          
        
    }
    getArgs () {
        const args = {};
        process.argv
            .slice(2, process.argv.length)
            .forEach( arg => {
            // long arg
            if (arg.slice(0,2) === '--') {
                const longArg = arg.split('=');
                const longArgFlag = longArg[0].slice(2,longArg[0].length);
                const longArgValue = longArg.length > 1 ? longArg[1] : true;
                args[longArgFlag] = longArgValue;
            }
            // flags
            else if (arg[0] === '-') {
                const flags = arg.slice(1,arg.length).split('');
                flags.forEach(flag => {
                args[flag] = true;
                });
            }
        });
        return args;
      }
}
