// cloud-firestore/dist/main/storage/storage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Storage = void 0;
const firebase_admin_1 = __importDefault(require("firebase-admin"));
const rxjs_1 = require("rxjs");
const transaction_1 = require("../firestore/transaction");
class Storage {
    constructor() {
        this.observer = new rxjs_1.Observable(observer => {
            observer.next();
            observer.complete();
        });
    }
    onInit(config) {
        if (!config)
            return;
        let adminConfig = {
            credential: firebase_admin_1.default.credential.cert(config.credential),
            storageBucket: config.server
        };
        if (config.appID) {
            firebase_admin_1.default.initializeApp(adminConfig, config.appID);
            let app = firebase_admin_1.default.app(config.appID);
            this.database = firebase_admin_1.default.firestore(app);
        }
        else {
            throw new Error("AppID required.");
        }
    }
    connect(isTransaction, currentTrans) {
        let observable = new rxjs_1.Observable(observer => {
            if ((!isTransaction && (currentTrans && currentTrans.type === transaction_1.Transaction.NO_TRANSACTION))
                || (currentTrans && currentTrans.type === transaction_1.Transaction.REQUIRED_TRANSACTION)) {
                observer.next(currentTrans);
                observer.complete();
            }
            else {
                if (isTransaction) {
                    observer.next(new transaction_1.Transaction({}, transaction_1.Transaction.REQUIRED_TRANSACTION));
                    observer.complete();
                }
                else {
                    observer.next(new transaction_1.Transaction({}, transaction_1.Transaction.NO_TRANSACTION));
                    observer.complete();
                }
            }
        });
        return observable;
    }
    commit(transaction) {
        if (transaction)
            return transaction.commit();
    }
    rollback(transaction) {
        if (transaction)
            return transaction.rollback();
    }
    post(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
            let file = input.files[0];
            var bucket = firebase_admin_1.default.storage().bucket();
            bucket.upload(file.name, { destination: file.name }).then((data) => {
                let file = data[0];
                observer.next([{ downloadUrl: "https://firebasestorage.googleapis.com/v0/b/" + bucket.name + "/o/" + encodeURIComponent(file.name) + "?alt=media&token=" + this.uuidv4() }, transaction]);
                observer.complete();
            }).catch(err => {
                observer.error(err);
            });
        });
    }
    uuidv4() {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }
    put(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
        });
    }
    delete(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
        });
    }
    get(input, pConfig, transaction) {
        return new rxjs_1.Observable(observer => {
        });
    }
}
exports.Storage = Storage;
