// quiver-component/out/lib/util/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _quiverObject = require("quiver-object");

var copy = _quiverObject.copy;
var noCopy = _quiverObject.noCopy;

const configNormalized = Symbol("ConfigNormalized");
const handlerMap = Symbol("handlerMap");
const bundleMap = Symbol("bundleMap");
const initTable = Symbol("initTable");

const globalConfig = function (config) {
  if (!config.global) {
    config.global = noCopy({});
  }

  return config.global;
};

exports.globalConfig = globalConfig;
const normalizeConfig = function (config) {
  const global = globalConfig(config);

  if (global[configNormalized]) return;

  if (!global[handlerMap]) {
    global[handlerMap] = noCopy({});
  }

  if (!global[bundleMap]) {
    global[bundleMap] = noCopy({});
  }

  if (!global[initTable]) {
    global[initTable] = {};
  }

  global[configNormalized] = true;

  return config;
};

exports.normalizeConfig = normalizeConfig;
const getHandlerMap = function (config) {
  normalizeConfig(config);

  return config.global[handlerMap];
};

exports.getHandlerMap = getHandlerMap;
const getBundleMap = function (config) {
  normalizeConfig(config);

  return config.global[bundleMap];
};

exports.getBundleMap = getBundleMap;
const getInitTable = function (config) {
  normalizeConfig(config);

  return config.global[initTable];
};

exports.getInitTable = getInitTable;
const getHandleable = function (config, component) {
  const handlerMap = getHandlerMap(config);
  return handlerMap[component.id];
};

exports.getHandleable = getHandleable;
const getBundle = function (config, component) {
  const bundleMap = getBundleMap(config);
  return bundleMap[component.id];
};
exports.getBundle = getBundle;
