// quiver-promise/out/lib/promise.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var error = require("quiver-error").error;

const createPromise = function (construct) {
  return new Promise(construct);
};

exports.createPromise = createPromise;
const resolve = function (val) {
  return Promise.resolve(val);
};

exports.resolve = resolve;
const reject = function (err) {
  return Promise.reject(err);
};

exports.reject = reject;
const timeout = function (time) {
  return createPromise(function (resolve) {
    return setTimeout(resolve, time);
  });
};

exports.timeout = timeout;
const safePromised = function (fn) {
  return function () {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return createPromise(function (resolve) {
      return resolve(fn.apply(undefined, args));
    });
  };
};

exports.safePromised = safePromised;
const promisify = function (fn) {
  return function () {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return createPromise(function (resolve, reject) {
      return fn.apply(undefined, args.concat([function (err, result) {
        return err ? reject(err) : resolve(result);
      }]));
    });
  };
};

exports.promisify = promisify;
const runAsync = function (gen) {
  const doNext = function (action) {
    let done, value;
    try {
      var _ref = action();

      done = _ref.done;
      value = _ref.value;
    } catch (err) {
      return reject(err);
    }

    if (done) return value;

    return resolve(value).then(function (result) {
      return doNext(function () {
        return gen.next(result);
      });
    }, function (err) {
      return doNext(function () {
        return gen["throw"](err);
      });
    });
  };

  return resolve(doNext(function () {
    return gen.next();
  }));
};

exports.runAsync = runAsync;
const async = function (fn) {
  return function () {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return resolve(fn.apply(undefined, args)).then(runAsync);
  };
};

exports.async = async;
const promisifyMethod = function (object, method) {
  return promisify(function () {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return object[method].apply(object, args);
  });
};

exports.promisifyMethod = promisifyMethod;
const promisifyMethods = function (object, methods) {
  return methods.reduce(function (result, method) {
    result[method] = promisifyMethod(object, method);
    return result;
  }, Object.create(object));
};
exports.promisifyMethods = promisifyMethods;
