// quiver-error/out/lib/error.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});
let error = function (errorCode, errorMessage, details) {
  return new ServerError(errorCode, errorMessage, details);
};

exports.error = error;

let ServerError = (function (_Error) {
  function ServerError() {
    let errorCode = arguments[0] === undefined ? 500 : arguments[0];
    let errorMessage = arguments[1] === undefined ? "" : arguments[1];
    let details = arguments[2] === undefined ? {} : arguments[2];

    _classCallCheck(this, ServerError);

    _get(Object.getPrototypeOf(ServerError.prototype), "constructor", this).call(this, errorMessage);

    Error.captureStackTrace(this, error);
    this._errorCode = errorCode;
    this._errorMessage = errorMessage;
    this._errorDetails = details;
  }

  _inherits(ServerError, _Error);

  _createClass(ServerError, {
    name: {
      get: function () {
        return "ServerError";
      }
    },
    code: {
      get: function () {
        return this._errorCode;
      }
    },
    errorCode: {
      get: function () {
        return this._errorCode;
      }
    },
    message: {
      get: function () {
        return this._errorMessage;
      }
    },
    details: {
      get: function () {
        return this._errorDetails;
      }
    },
    format: {
      value: function format() {
        return "[Error " + this.errorCode + ": " + this.message + "]\n" + this.stack;
      }
    },
    inspect: {
      value: function inspect() {
        return this.format();
      }
    }
  });

  return ServerError;
})(Error);
