// quiver-http/out/lib/handler.js
"use strict";

var _slicedToArray = function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { var _arr = []; for (var _iterator = arr[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) { _arr.push(_step.value); if (i && _arr.length === i) break; } return _arr; } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var async = require("quiver-promise").async;

var normalizeHttpHeader = require("./normalize").normalizeHttpHeader;

var _header = require("./header");

var RequestHead = _header.RequestHead;
var ResponseHead = _header.ResponseHead;

var _quiverStreamUtil = require("quiver-stream-util");

var pipeStream = _quiverStreamUtil.pipeStream;
var streamToStreamable = _quiverStreamUtil.streamToStreamable;
var nodeToQuiverReadStream = _quiverStreamUtil.nodeToQuiverReadStream;
var nodeToQuiverWriteStream = _quiverStreamUtil.nodeToQuiverWriteStream;
const streamToHttpHandler = function (streamHandler) {
  return function (requestHead, requestStreamable) {
    const args = requestHead.args;

    const contentType = requestHead.getHeader("content-type");
    const contentLength = requestHead.getHeader("content-length");

    if (contentType) requestStreamable.contentType = contentType;

    if (contentLength) requestStreamable.contentLength = parseInt(contentLength);

    return streamHandler(args, requestStreamable).then(function (resultStreamable) {
      const responseHead = new ResponseHead();

      const contentType = resultStreamable.contentType;
      const contentLength = resultStreamable.contentLength;

      if (contentType) {
        responseHead.setHeader("content-type", contentType);
      }

      if (typeof contentLength == "number") {
        responseHead.setHeader("content-length", (contentLength | 0).toString());
      }

      return [responseHead, resultStreamable];
    });
  };
};

exports.streamToHttpHandler = streamToHttpHandler;
const httpToNodeHandler = function (httpHandler) {
  return async(function* (request, response) {
    try {
      const requestHead = new RequestHead({
        httpVersion: request.httpVersion,
        headers: request.headers,
        method: request.method,
        url: request.url
      });

      requestHead.setArgs("clientAddress", request.connection.remoteAddress);

      const requestStreamable = streamToStreamable(nodeToQuiverReadStream(request));

      var _ref = yield httpHandler(requestHead, requestStreamable);

      var _ref2 = _slicedToArray(_ref, 2);

      const responseHead = _ref2[0];
      const responseStreamable = _ref2[1];

      const headers = responseHead.headers;

      for (let key in headers) {
        const normalizedKey = normalizeHttpHeader(key, true);
        response.setHeader(normalizedKey, headers[key]);
      }

      response.writeHead(responseHead.statusCode, responseHead.statusMessage);

      if (headers["transfer-encoding"] == "chunked") {
        response.chunkedEncoding = false;
      }

      if (responseStreamable.toNodeStream) {
        const nodeRead = yield responseStreamable.toNodeStream();
        nodeRead.pipe(response);
      } else {
        const responseStream = yield responseStreamable.toStream();
        const responseWrite = nodeToQuiverWriteStream(response);

        pipeStream(responseStream, responseWrite);
      }
    } catch (err) {
      if (!response.headersSents) {
        const errorCode = err.errorCode || 500;

        response.writeHead(errorCode, {
          "Content-Length": 0
        });
      }

      response.end();
    }
  });
};
exports.httpToNodeHandler = httpToNodeHandler;
