// quiver-http/out/lib/header.js
"use strict";

var _interopRequire = function (obj) { return obj && obj.__esModule ? obj["default"] : obj; };

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var http = _interopRequire(require("http"));

var urlLib = _interopRequire(require("url"));

var copy = require("quiver-object").copy;

const parseUrl = urlLib.parse;
const formatUrl = urlLib.format;

var qs = _interopRequire(require("querystring"));

const parseQueryString = qs.parse;
const queryStringify = qs.stringify;

const statusCodeTable = http.STATUS_CODES;

const getStatusMessage = function (statusCode) {
  return statusCodeTable[statusCode] || "Unknown";
};

const assertString = function (str) {
  if (typeof str != "string") throw new Error("argument must be string");

  return str;
};

const assertNumber = function (num) {
  if (typeof num != "number") throw new Error("argument must be number");

  return num;
};

const assertRegex = function (str, regex) {
  if (regex.test(assertString(str))) throw new Error("string contains invalid characters");

  return str;
};

const fieldRegex = /[^a-zA-Z\-]/;
const valueRegex = /[^\x20-\x7E]/;
const methodRegex = /[^a-zA-Z]/;

let HttpHead = (function () {
  function HttpHead() {
    let rawHead = arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, HttpHead);

    var _rawHead$httpVersion = rawHead.httpVersion;
    const httpVersion = _rawHead$httpVersion === undefined ? "1.1" : _rawHead$httpVersion;

    this._httpVersion = httpVersion;

    if (rawHead.headers) {
      const rawHeaders = rawHead.headers;

      this._headers = Object.keys(rawHeaders).reduce(function (headers, key) {
        headers[key.toLowerCase()] = assertString(rawHeaders[key]);

        return headers;
      }, {});
    } else {
      this._headers = {};
    }
  }

  _createClass(HttpHead, {
    httpVersion: {
      get: function () {
        return this._httpVersion;
      }
    },
    setHeader: {
      value: function setHeader(header, value) {
        const key = assertRegex(header, fieldRegex).toLowerCase();

        this._headers[key] = assertRegex(value, valueRegex);

        return this;
      }
    },
    getHeader: {
      value: function getHeader(header) {
        return this._headers[header.toLowerCase()];
      }
    },
    removeHeader: {
      value: function removeHeader(header) {
        delete this._headers[header.toLowerCase()];
      }
    },
    headers: {
      get: function () {
        return this._headers;
      }
    }
  });

  return HttpHead;
})();

let RequestHead = exports.RequestHead = (function (_HttpHead) {
  function RequestHead() {
    let rawHead = arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, RequestHead);

    var _rawHead$method = rawHead.method;
    const method = _rawHead$method === undefined ? "GET" : _rawHead$method;
    var _rawHead$url = rawHead.url;
    const url = _rawHead$url === undefined ? "/" : _rawHead$url;

    this._method = assertRegex(method, methodRegex);

    this._url = assertString(url);
    this._args = {};

    this._reset();

    _get(Object.getPrototypeOf(RequestHead.prototype), "constructor", this).call(this, rawHead);

    this._args.path = this.path, this._args.requestHead = this;
  }

  _inherits(RequestHead, _HttpHead);

  _createClass(RequestHead, {
    method: {
      get: function () {
        return this._method;
      },
      set: function (newMethod) {
        this._method = assertRegex(newMethod, methodRegex).toUpperCase();
      }
    },
    _reset: {
      value: function _reset() {
        this._modifiedUrl = false;
        this._modifiedQuery = false;

        this._path = null;
        this._protocol = null;
        this._hostname = null;
        this._port = null;
        this._auth = null;

        this.__parsedUrl = null;
        this._queryString = null;
        this.__parsedQuery = null;
      }
    },
    reformatUrl: {
      value: function reformatUrl() {
        const url = this._url = formatUrl({
          pathname: this.path,
          search: this.queryString,
          protocol: this.protocol,
          hostname: this.hostname,
          port: this.port,
          auth: this.auth
        });

        this._modifiedUrl = false;
        this.__parsedUrl = null;

        return url;
      }
    },
    url: {
      get: function () {
        if (!this._modifiedUrl) return this._url;

        return this.reformatUrl();
      },
      set: function (newUrl) {
        this._url = newUrl;
        this._reset();
      }
    },
    _parsedUrl: {
      get: function () {
        if (this.__parsedUrl) return this.__parsedUrl;

        const parsed = this.__parsedUrl = parseUrl(this._url);
        return parsed;
      }
    },
    urlComponents: {
      get: function () {
        if (this._modifiedUrl) this.reformatUrl();

        return copy(this._parsedUrl);
      }
    },
    path: {
      get: function () {
        if (this._path) return this._path;

        return this.args.path = this._path = this._parsedUrl.pathname;
      },
      set: function (newPath) {
        this._path = assertString(newPath);
        this.args.path = newPath;
        this._modifiedUrl = true;
      }
    },
    queryString: {
      get: function () {
        if (this._modifiedQuery) {
          const queryString = this._queryString = queryStringify(this.__parsedQuery);

          this._modifiedQuery = false;
          return queryString;
        }

        if (this._queryString) {
          return this._queryString;
        }

        const queryString = this._queryString = this._parsedUrl.query || "";

        return queryString;
      },
      set: function (str) {
        this._queryString = assertString(str);

        this.__parsedQuery = null;
        this._modifiedUrl = true;
        this._modifiedQuery = false;
      }
    },
    _parsedQuery: {
      get: function () {
        if (this.__parsedQuery) return this.__parsedQuery;

        const query = this.__parsedQuery = parseQueryString(this.queryString);

        return query;
      }
    },
    query: {
      get: function () {
        return copy(this._parsedQuery);
      }
    },
    setQuery: {
      value: function setQuery(key, value) {
        this._parsedQuery[assertString(key)] = assertString(value);

        this._modifiedQuery = true;
        this._modifiedUrl = true;

        return this;
      }
    },
    setArgs: {
      value: function setArgs(key, value) {
        this._args[key] = value;

        return this;
      }
    },
    args: {
      get: function () {
        return this._args;
      }
    }
  });

  return RequestHead;
})(HttpHead);

const mixinUrlComponent = function (prototype, field) {
  let urlField = arguments[2] === undefined ? field : arguments[2];
  let defaultValue = arguments[3] === undefined ? null : arguments[3];
  return (function () {
    const _field = "_" + field;

    Object.defineProperty(prototype, field, {
      get: function get() {
        if (this[_field] !== null) {
          return this[_field];
        }const value = this[_field] = this._parsedUrl[urlField] || defaultValue;

        return value;
      },

      set: function set(value) {
        this[_field] = value;
        this._modifiedUrl = true;

        return this;
      }
    });
  })();
};

mixinUrlComponent(RequestHead.prototype, "protocol");
mixinUrlComponent(RequestHead.prototype, "hostname");
mixinUrlComponent(RequestHead.prototype, "port");
mixinUrlComponent(RequestHead.prototype, "auth");

let ResponseHead = exports.ResponseHead = (function (_HttpHead2) {
  function ResponseHead() {
    let rawHead = arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, ResponseHead);

    var _rawHead$statusCode = rawHead.statusCode;
    const statusCode = _rawHead$statusCode === undefined ? 200 : _rawHead$statusCode;
    var _rawHead$statusMessage = rawHead.statusMessage;
    const statusMessage = _rawHead$statusMessage === undefined ? getStatusMessage(statusCode) : _rawHead$statusMessage;

    this._statusCode = assertNumber(statusCode);
    this._statusMessage = assertRegex(statusMessage, valueRegex);

    _get(Object.getPrototypeOf(ResponseHead.prototype), "constructor", this).call(this, rawHead);
  }

  _inherits(ResponseHead, _HttpHead2);

  _createClass(ResponseHead, {
    statusCode: {
      get: function () {
        return this._statusCode;
      },
      set: function (code) {
        this._statusCode = assertNumber(code);
        this._statusMessage = getStatusMessage(code);
      }
    },
    statusMessage: {
      get: function () {
        return this._statusMessage;
      },
      set: function (message) {
        this._statusMessage = assertRegex(message, valueRegex);
      }
    }
  });

  return ResponseHead;
})(HttpHead);
