// quiver-http/out/lib/subrequest.js
"use strict";

var _interopRequire = function (obj) { return obj && obj.__esModule ? obj["default"] : obj; };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var http = _interopRequire(require("http"));

var _header = require("./header");

var RequestHead = _header.RequestHead;
var ResponseHead = _header.ResponseHead;

var _quiverStreamUtil = require("quiver-stream-util");

var pipeStream = _quiverStreamUtil.pipeStream;
var streamToStreamable = _quiverStreamUtil.streamToStreamable;
var nodeToQuiverReadStream = _quiverStreamUtil.nodeToQuiverReadStream;
var nodeToQuiverWriteStream = _quiverStreamUtil.nodeToQuiverWriteStream;
const requestHeadToOptions = function (requestHead) {
  const auth = requestHead.auth;
  const hostname = requestHead.hostname;
  const port = requestHead.port;
  const method = requestHead.method;
  const path = requestHead.path;
  const queryString = requestHead.queryString;
  const headers = requestHead.headers;

  return {
    auth: auth,
    hostname: hostname,
    port: port,
    method: method,
    path: path + "?" + queryString,
    headers: headers
  };
};

exports.requestHeadToOptions = requestHeadToOptions;
const nodeToQuiverResponse = function (response) {
  const responseHead = new ResponseHead(response);

  const responseStream = nodeToQuiverReadStream(response);

  return [responseHead, responseStream];
};

exports.nodeToQuiverResponse = nodeToQuiverResponse;
const subrequest = function (requestHead, requestStream) {
  return new Promise(function (resolve, reject) {
    const request = http.request(requestHeadToOptions(requestHead), function (response) {
      return resolve(nodeToQuiverResponse(response));
    }).on("error", reject);

    pipeStream(requestStream, nodeToQuiverWriteStream(request));
  });
};

exports.subrequest = subrequest;
const getRequest = function (url, requestStream) {
  return new Promise(function (resolve, reject) {
    http.get(url, function (response) {
      return resolve(nodeToQuiverResponse(response));
    }).on("error", reject);
  });
};
exports.getRequest = getRequest;
