// quiver-stream-channel/out/lib/promised.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var createPromise = require("quiver-promise").createPromise;

var simpleChannel = require("./simple").simpleChannel;

const promisedReadStream = function (simpleReadStream) {
  return {
    read: function () {
      return createPromise(function (resolve, reject) {
        return simpleReadStream.read(function (readClosed, data) {
          if (!readClosed) {
            resolve({ data: data });
          } else if (!readClosed.err) {
            resolve({ closed: true });
          } else {
            reject(readClosed.err);
          }
        });
      });
    },

    closeRead: simpleReadStream.closeRead,
    isClosed: simpleReadStream.isClosed
  };
};

exports.promisedReadStream = promisedReadStream;
const promisedWriteStream = function (simpleWriteStream) {
  return {
    prepareWrite: function () {
      return createPromise(function (resolve, reject) {
        return simpleWriteStream.prepareWrite(function (writeClosed) {
          if (!writeClosed) {
            resolve({});
          } else if (!writeClosed.err) {
            resolve({ closed: true });
          } else {
            reject(writeClosed.err);
          }
        });
      });
    },

    write: simpleWriteStream.write,
    closeWrite: simpleWriteStream.closeWrite,
    isClosed: simpleWriteStream.isClosed
  };
};

exports.promisedWriteStream = promisedWriteStream;
const promisedChannel = function () {
  var _simpleChannel = simpleChannel();

  const simpleRead = _simpleChannel.readStream;
  const simpleWrite = _simpleChannel.writeStream;

  const readStream = promisedReadStream(simpleRead);
  const writeStream = promisedWriteStream(simpleWrite);

  return { readStream: readStream, writeStream: writeStream };
};
exports.promisedChannel = promisedChannel;
