// cloud-firestore/dist/test/test-sql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestSql = void 0;
const store_1 = require("../main/firestore/store");
const util_1 = require("./util");
//import express from 'express';
//import bodyParser from 'body-parser';
const patient_loader_1 = require("./patient-loader");
class TestSql {
    constructor() {
        this.provider = new store_1.Store();
        this.emrLoader = new patient_loader_1.PatientLoader();
    }
    load() {
        this.emrLoader.load(this.provider);
    }
    createServer() {
        /*const app = express();
        app.use(bodyParser.json({
            limit: '50mb'
          }));
      
          app.use(bodyParser.urlencoded({
            limit: '50mb',
            parameterLimit: 100000,
            extended: true
          }));
          let args = this.getArgs();
          var serverPort = process.env.PORT || args['port'];
          app.listen(serverPort, ()=>{  return console.log(`server is listening on ${serverPort}`);
          });
          this.query(app);*/
    }
    query(app) {
        let sConfig = { 'serviceId': "TM_APPOINTMENTS", 'input': {}, 'output': { 'isPassthrough': true }, 'oprType': {}, 'target': 'adminStore', Util: util_1.Util };
        app.get('/load', (req, res) => {
            this.provider.connect(true, null).subscribe((cmd) => {
                //let reqObj:any = {filter: "DOCTCODE eq '000001' and ((APPOINTMENTDATE ge 1598832000000) and APPOINTMENTDATE le 1600905600000)"};//{"filter":"mobile_code ne null"};            
                let reqObj = { select: "VISITSTATUS,DOCTCODE,APPOINTMENTDATE,APPOINTMENTNO,MOBILE", filter: "DOCTCODE eq '000001' and ((APPOINTMENTDATE ge '2020-08-01') and APPOINTMENTDATE le '2020-11-01')" };
                this.provider.get({ body: reqObj, headers: {} }, sConfig, cmd).subscribe(result => {
                    cmd.commit();
                    res.send(JSON.stringify(result[0]));
                    res.status(200);
                    res.end();
                });
            });
        });
    }
    getArgs() {
        const args = {};
        process.argv
            .slice(2, process.argv.length)
            .forEach(arg => {
            // long arg
            if (arg.slice(0, 2) === '--') {
                const longArg = arg.split('=');
                const longArgFlag = longArg[0].slice(2, longArg[0].length);
                const longArgValue = longArg.length > 1 ? longArg[1] : true;
                args[longArgFlag] = longArgValue;
            }
            // flags
            else if (arg[0] === '-') {
                const flags = arg.slice(1, arg.length).split('');
                flags.forEach(flag => {
                    args[flag] = true;
                });
            }
        });
        return args;
    }
}
exports.TestSql = TestSql;
