// cloud-firestore/dist/test/uploader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Uploader = void 0;
const store_1 = require("../main/firestore/store");
const util_1 = require("./util");
const patient_loader_1 = require("./patient-loader");
class Uploader {
    constructor() {
        this.provider = new store_1.Store();
        this.emrLoader = new patient_loader_1.PatientLoader();
    }
    load() {
        this.emrLoader.load(this.provider);
    }
    save() {
        let app = null;
        /*const app = express();
        app.use(bodyParser.json({
            limit: '50mb'
          }));
      
          app.use(bodyParser.urlencoded({
            limit: '50mb',
            parameterLimit: 100000,
            extended: true
          }));
          let args = this.getArgs();
          var serverPort = process.env.PORT || args['port'];
          app.listen(serverPort, ()=>{  return console.log(`server is listening on ${serverPort}`);
          });*/
        let sConfig = { 'serviceId': "MST_COUNTRY", 'input': {}, 'output': { 'isPassthrough': true }, 'oprType': {}, 'target': 'adminStore', Util: util_1.Util };
        app.get('/upload', (req, res) => {
            let counter = 0;
            let tableId = req.query.table;
            sConfig.serviceId = tableId;
            debugger;
            this.provider.connect(true, null).subscribe((cmd) => {
                let items = this.emrLoader[tableId];
                items.forEach((item) => {
                    if (item.IMAGE)
                        delete item.IMAGE;
                    Object.keys(item).forEach((element) => {
                        let data = item[element];
                        if (typeof data === 'string' && this.isValidDate(data)) {
                            item[element] = new Date(data);
                        }
                    });
                    let reqObj = { insert: item };
                    let request = { body: reqObj, headers: req.headers };
                    this.provider.post(request, sConfig, cmd).subscribe(result => {
                        counter++;
                        if (counter === items.length) {
                            cmd.commit();
                            res.send(JSON.stringify(result[0]));
                            res.status(200);
                            res.end();
                        }
                    });
                });
            });
        });
        app.get('/load', (req, res) => {
            this.provider.connect(true, null).subscribe((cmd) => {
                let reqObj = {}; //{"filter":"mobile_code ne null"};            
                this.provider.get({ body: reqObj, headers: {} }, sConfig, cmd).subscribe(result => {
                    //cmd.commit();
                    res.send(JSON.stringify(result[0]));
                    res.status(200);
                    res.end();
                });
            });
        });
    }
    isValidDate(dateString) {
        var regEx = /^\d{4}-\d{2}-\d{2}$/;
        if (!dateString.match(regEx))
            return false; // Invalid format
        var d = new Date(dateString);
        var dNum = d.getTime();
        if (!dNum && dNum !== 0)
            return false; // NaN value, Invalid date
        return d.toISOString().slice(0, 10) === dateString;
    }
    getArgs() {
        const args = {};
        process.argv
            .slice(2, process.argv.length)
            .forEach(arg => {
            // long arg
            if (arg.slice(0, 2) === '--') {
                const longArg = arg.split('=');
                const longArgFlag = longArg[0].slice(2, longArg[0].length);
                const longArgValue = longArg.length > 1 ? longArg[1] : true;
                args[longArgFlag] = longArgValue;
            }
            // flags
            else if (arg[0] === '-') {
                const flags = arg.slice(1, arg.length).split('');
                flags.forEach(flag => {
                    args[flag] = true;
                });
            }
        });
        return args;
    }
}
exports.Uploader = Uploader;
