// quiver-component/out/lib/composite/list.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var Component = require("../component").Component;

const _componentArray = Symbol("@componentArray");

let ListComponent = exports.ListComponent = (function (_Component) {
  function ListComponent() {
    let componentList = arguments[0] === undefined ? [] : arguments[0];
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, ListComponent);

    _get(Object.getPrototypeOf(ListComponent.prototype), "constructor", this).call(this, options);
    this[_componentArray] = componentList;
  }

  _inherits(ListComponent, _Component);

  _createClass(ListComponent, {
    each: {
      value: function each(iteratee) {
        this[_componentArray].forEach(iteratee);
        _get(Object.getPrototypeOf(ListComponent.prototype), "each", this).call(this, iteratee);
      }
    },
    doMap: {
      value: function doMap(target, mapper, mapTable) {
        target[_componentArray] = this[_componentArray].map(function (component) {
          return component.applyMap(mapper, mapTable);
        });

        _get(Object.getPrototypeOf(ListComponent.prototype), "doMap", this).call(this, target, mapper, mapTable);
      }
    },
    push: {
      value: function push(component) {
        if (!component.isQuiverComponent) {
          throw new TypeError("object must be a quiver component");
        }

        this[_componentArray].push(component);
      }
    },
    unshift: {
      value: function unshift(component) {
        if (!component.isQuiverComponent) {
          throw new TypeError("object must be a quiver component");
        }

        this[_componentArray].unshift(component);
      }
    },
    array: {
      get: function () {
        return this[_componentArray];
      }
    }
  });

  return ListComponent;
})(Component);

const listComponent = function () {
  let components = arguments[0] === undefined ? [] : arguments[0];
  return new ListComponent(components);
};
exports.listComponent = listComponent;
