// quiver-component/out/lib/composite/map.js
"use strict";

var _slicedToArray = function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { var _arr = []; for (var _iterator = arr[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) { _arr.push(_step.value); if (i && _arr.length === i) break; } return _arr; } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } };

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

Object.defineProperty(exports, "__esModule", {
  value: true
});

var Component = require("../component").Component;

const _componentMap = Symbol("@componentMap");

let MapComponent = exports.MapComponent = (function (_Component) {
  function MapComponent() {
    let componentMap = arguments[0] === undefined ? new Map() : arguments[0];
    let options = arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, MapComponent);

    _get(Object.getPrototypeOf(MapComponent.prototype), "constructor", this).call(this, options);
    this[_componentMap] = componentMap;
  }

  _inherits(MapComponent, _Component);

  _createClass(MapComponent, {
    each: {
      value: function each(iteratee) {
        for (let component of this[_componentMap].values()) {
          iteratee(component);
        }

        _get(Object.getPrototypeOf(MapComponent.prototype), "each", this).call(this, iteratee);
      }
    },
    doMap: {
      value: function doMap(target, mapper, mapTable) {
        const targetMap = new Map();

        for (let _ref of this[_componentMap].entries()) {
          var _ref2 = _slicedToArray(_ref, 2);

          let key = _ref2[0];
          let component = _ref2[1];

          targetMap[key] = component.applyMap(mapper, mapTable);
        }

        allKeys(componentMap).forEach(function (key) {
          const component = componentMap[key];
          targetMap[key] = component.applyMap(mapper, mapTable);
        });

        target[_componentMap] = targetMap;
        _get(Object.getPrototypeOf(MapComponent.prototype), "doMap", this).call(this, target, mapper, mapTable);
      }
    },
    set: {
      value: function set(key, component) {
        if (!component.isQuiverComponent) {
          throw new TypeError("object must be a quiver component");
        }

        this[_componentMap].set(key, component);
      }
    },
    get: {
      value: function get(key) {
        return this[_componentMap].get(key);
      }
    },
    map: {
      get: function () {
        return this[_componentMap];
      }
    }
  });

  return MapComponent;
})(Component);

const mapComponent = function () {
  let components = arguments[0] === undefined ? new Map() : arguments[0];
  return new MapComponent(components);
};
exports.mapComponent = mapComponent;
